{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Filter projected genes \n",
    "\n",
    "There are two different sets of gene set. One for the assemblies in pseudomolecules (PGSBv2: ```arinalrfor, chinese, jagger, julius, lancer, landmark, mace, norin61, spelta, stanley, sy_mattis```) and in scaffolds (EI: ```cadenza, claire, paragon, robigus, weebil```). \n",
    "\n",
    "## EI annotations\n",
    "\n",
    "The source gene from the RefSeq v1.1 (IWGSC, 2018) was obtained from ```UK.geneid2CSinformant.txt```, corresponding to the mapping is described in 10+ Wheat Genomes Consortium (2020). Each scaffold was classified according to the corresponding chromosome from the source genes, excluding projections classified as pseudogenes. Then each gene was classified as being consistent or not with the expected scaffold. F\n",
    "\n",
    "\n",
    "## PGSBv2  annotation\n",
    "\n",
    "The projected genes where classified as being consistent between the chromosome from the source and projected genes.   \n",
    "\n",
    "## Both annotations.\n",
    "\n",
    "Then, the genes on each annotation where filtered to keep the genes that map to the expected chromosome. Finally, the genes that have more than one projection in the expected chromosome where removed.  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "if(window['d3'] === undefined ||\n",
       "   window['Nyaplot'] === undefined){\n",
       "    var path = {\"d3\":\"https://cdnjs.cloudflare.com/ajax/libs/d3/3.5.5/d3.min\",\"downloadable\":\"http://cdn.rawgit.com/domitry/d3-downloadable/master/d3-downloadable\"};\n",
       "\n",
       "\n",
       "\n",
       "    var shim = {\"d3\":{\"exports\":\"d3\"},\"downloadable\":{\"exports\":\"downloadable\"}};\n",
       "\n",
       "    require.config({paths: path, shim:shim});\n",
       "\n",
       "\n",
       "require(['d3'], function(d3){window['d3']=d3;console.log('finished loading d3');require(['downloadable'], function(downloadable){window['downloadable']=downloadable;console.log('finished loading downloadable');\n",
       "\n",
       "\tvar script = d3.select(\"head\")\n",
       "\t    .append(\"script\")\n",
       "\t    .attr(\"src\", \"http://cdn.rawgit.com/domitry/Nyaplotjs/master/release/nyaplot.js\")\n",
       "\t    .attr(\"async\", true);\n",
       "\n",
       "\tscript[0][0].onload = script[0][0].onreadystatechange = function(){\n",
       "\n",
       "\n",
       "\t    var event = document.createEvent(\"HTMLEvents\");\n",
       "\t    event.initEvent(\"load_nyaplot\",false,false);\n",
       "\t    window.dispatchEvent(event);\n",
       "\t    console.log('Finished loading Nyaplotjs');\n",
       "\n",
       "\t};\n",
       "\n",
       "\n",
       "});});\n",
       "}\n"
      ],
      "text/plain": [
       "\"if(window['d3'] === undefined ||\\n   window['Nyaplot'] === undefined){\\n    var path = {\\\"d3\\\":\\\"https://cdnjs.cloudflare.com/ajax/libs/d3/3.5.5/d3.min\\\",\\\"downloadable\\\":\\\"http://cdn.rawgit.com/domitry/d3-downloadable/master/d3-downloadable\\\"};\\n\\n\\n\\n    var shim = {\\\"d3\\\":{\\\"exports\\\":\\\"d3\\\"},\\\"downloadable\\\":{\\\"exports\\\":\\\"downloadable\\\"}};\\n\\n    require.config({paths: path, shim:shim});\\n\\n\\nrequire(['d3'], function(d3){window['d3']=d3;console.log('finished loading d3');require(['downloadable'], function(downloadable){window['downloadable']=downloadable;console.log('finished loading downloadable');\\n\\n\\tvar script = d3.select(\\\"head\\\")\\n\\t    .append(\\\"script\\\")\\n\\t    .attr(\\\"src\\\", \\\"http://cdn.rawgit.com/domitry/Nyaplotjs/master/release/nyaplot.js\\\")\\n\\t    .attr(\\\"async\\\", true);\\n\\n\\tscript[0][0].onload = script[0][0].onreadystatechange = function(){\\n\\n\\n\\t    var event = document.createEvent(\\\"HTMLEvents\\\");\\n\\t    event.initEvent(\\\"load_nyaplot\\\",false,false);\\n\\t    window.dispatchEvent(event);\\n\\t    console.log('Finished loading Nyaplotjs');\\n\\n\\t};\\n\\n\\n});});\\n}\\n\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "false"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "require 'zlib'\n",
    "require 'bio'\n",
    "require 'csv'\n",
    "require 'set'\n",
    "require 'bio-gff3'\n",
    "require 'daru'\n",
    "require 'nyaplot'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       ":parseEITranscript"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Transcript = Struct.new(:id, :gene, :chromosome,:version,:count,:transcript,:confidence, :count_int, :isoform)\n",
    "def parseTranscript name\n",
    "    arr=name.split(\".\")\n",
    "    match = /TraesCS(?<chr>[[:alnum:]]{1,2})(?<ver>[[:digit:]]{2})G(?<count>[[:digit:]]+)(?<conf>[[:upper:]]*)/.match arr[0]\n",
    "    raise \"Unable to parse: #{name}\" unless match\n",
    "    Transcript.new(name, arr[0],match[:chr],match[:ver],match[:count],arr[1],match[:conf], match[:count].to_i, arr[1])\n",
    "end\n",
    "\n",
    "def parseEITranscript name\n",
    "    arr=name.split(\".\")\n",
    "    match = /Traes(?<chr>[[:upper:]]{3}_scaffold_[[:digit:]]*)_(?<ver>[[:digit:]]{2})G(?<count>[[:digit:]]+)(?<conf>[[:upper:]]*)/.match arr[0]\n",
    "    raise \"Unable to parse: #{name}\" unless match\n",
    "    Transcript.new(name, arr[0],match[:chr].downcase,match[:ver],match[:count],arr[1],match[:conf], match[:count].to_i, arr[1])\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "610352"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MrnaStats = Struct.new(:cds_count, :cds_max_gap)\n",
    "module Bio::GFFbrowser::FastLineParser\n",
    "  module_function :parse_line_fast\n",
    "end\n",
    "\n",
    "class GFF3\n",
    "    def initialize(file: \"\", is_gz: true)\n",
    "        @file = file\n",
    "        @is_gz = is_gz\n",
    "    end\n",
    "    \n",
    "    def header\n",
    "        return enum_for(:header) unless block_given? \n",
    "        io = nil\n",
    "        if @is_gz\n",
    "            infile = open(@file)\n",
    "            io = Zlib::GzipReader.new(infile) \n",
    "        else\n",
    "            io =  File.open(@file)\n",
    "        end\n",
    "        \n",
    "        io.each_line do |line|  \n",
    "            line.encode!('UTF-8', 'UTF-8', :invalid => :replace)\n",
    "            line.strip!\n",
    "            break if line == '##FASTA'\n",
    "            break unless line.length == 0 or line =~ /^#/\n",
    "            yield line\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    def each\n",
    "        return enum_for(:each) unless block_given? \n",
    "        io = nil\n",
    "        if @is_gz\n",
    "            infile = open(@file)\n",
    "            io = Zlib::GzipReader.new(infile) \n",
    "        else\n",
    "            io =  File.open(@file)\n",
    "        end\n",
    "        parser = Bio::GFFbrowser::FastLineParser\n",
    "        io.each_line do |line|  \n",
    "            line.encode!('UTF-8', 'UTF-8', :invalid => :replace)\n",
    "            line.strip!\n",
    "            break if line == '##FASTA'\n",
    "            #parents.clear if line == '###'\n",
    "            next if line.length == 0 or line =~ /^#/\n",
    "            begin\n",
    "                record = Bio::GFFbrowser::FastLineRecord.new(parser.parse_line_fast(line))\n",
    "                yield record\n",
    "            rescue Exception => e\n",
    "                $stderr.puts \"Unable to parse '#{line}'\\n#{e}\" \n",
    "                throw e\n",
    "            end\n",
    "         end\n",
    "    end\n",
    "    \n",
    "    def each_gene\n",
    "        return enum_for(:each_gene) unless block_given? \n",
    "        self.each do |record|\n",
    "           next unless record.feature == \"gene\"\n",
    "            yield record\n",
    "        end \n",
    "    end\n",
    "    \n",
    "    def each_mrna\n",
    "        return enum_for(:each_mrna) unless block_given? \n",
    "        self.each do |record|\n",
    "           next unless record.feature == \"mRNA\"\n",
    "            yield record\n",
    "        end \n",
    "    end\n",
    "    \n",
    "    def each_cds\n",
    "        return enum_for(:each_mrna) unless block_given? \n",
    "        self.each do |record|\n",
    "           next unless record.feature == \"CDS\"\n",
    "            yield record\n",
    "        end \n",
    "    end\n",
    "    \n",
    "    def calculate_mrna_stats\n",
    "        return if @mrna_stats\n",
    "        \n",
    "        @mrna_stats = Hash.new {|h,k| h[k] = MrnaStats.new(0,0) }\n",
    "        \n",
    "        last_mrna = \"\"\n",
    "        last_record = nil\n",
    "        each_cds do |record|\n",
    "            parent = record.get_attribute \"Parent\"\n",
    "            mrna = @mrna_stats[parent]\n",
    "            mrna.cds_count += 1\n",
    "            \n",
    "            if last_mrna == parent\n",
    "                distance =  record.start - last_record.end \n",
    "                mrna.cds_max_gap = distance if distance > mrna.cds_max_gap\n",
    "            end\n",
    "            last_record = record\n",
    "            last_mrna   = parent\n",
    "\n",
    "        end\n",
    "        return\n",
    "    end\n",
    "    \n",
    "    def mrna_info(id)\n",
    "        calculate_mrna_stats\n",
    "        @mrna_stats[id] \n",
    "    end\n",
    "    \n",
    "    def cds_to_print(mrna,cannonical_exons:[])\n",
    "        cds_features = [] \n",
    "        i = 0\n",
    "        offset=0\n",
    "        offset_start=0\n",
    "        each_cds do |record|\n",
    "            # Target=Sm1_CDS 3049 3183 +\n",
    "            target = record.get_attribute \"Target\"\n",
    "            arr = target.split(\" \")\n",
    "            col = @@colors[i % @@colors.size ]\n",
    "            start = arr[1].to_i + offset\n",
    "            ends = arr[2].to_i + offset\n",
    "            offset_start = record.start  if offset_start == 0\n",
    "            tmp = CDS_feature.new(start, ends, col, \n",
    "                record.seqid, record.start,record.end, record.start - offset_start )\n",
    "            #offset = ends\n",
    "            cds_features << tmp\n",
    "            i += 1\n",
    "        end\n",
    "        cds_features\n",
    "    end\n",
    "    \n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "    \n",
    "class MultipleGFFs\n",
    "    def initialize(folder: \"../mapping/\", lines:[], suffix:\".SM1.cds.sorted.gff\", is_gz:false )\n",
    "        @folder = folder\n",
    "        @lines = lines\n",
    "        @suffix = suffix\n",
    "        @lines_gffs = Hash.new\n",
    "        @lines.each do |l|\n",
    "            path =\"#{folder}/#{l}#{suffix}\"\n",
    "            @lines_gffs[l] = GFF3.new(file: path, is_gz: is_gz)\n",
    "            #@lines_gffs[l].calculate_mrna_stats\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    def method_missing(m, *args, &block)\n",
    "        @lines_gffs.send(m, *args, &block)\n",
    "    end\n",
    "    \n",
    "    def each_gff\n",
    "       @lines_gffs.each_pair{|k,v| yield k, v }\n",
    "    end\n",
    "    \n",
    "    def summary\n",
    "        ret = []\n",
    "        each_gff do |k,v|\n",
    "            v.each_mrna do |record|\n",
    "                #coverage=100.0;identity=100.0;matches=4443;mismatches=0;indels=0;unknowns=0\n",
    "                tmp = {}\n",
    "                tmp[:line] = k\n",
    "                tmp[:id] = record.get_attribute \"Name\"\n",
    "                tmp[:chr] = record.seqid\n",
    "                tmp[:start] = record.start\n",
    "                tmp[:end] = record.end\n",
    "                tmp[:strand] = record.strand\n",
    "                tmp[:genomic_length] = record.end - record.start\n",
    "                tmp[:coverage] = record.get_attribute \"coverage\"\n",
    "                tmp[:identity] = record.get_attribute \"identity\"\n",
    "                tmp[:matches]  = record.get_attribute \"matches\"\n",
    "                tmp[:mismatches]  = record.get_attribute \"mismatches\"\n",
    "                tmp[:indels] = record.get_attribute \"indels\"\n",
    "                tmp[:unknowns] = record.get_attribute \"unknowns\"\n",
    "                mrna_stats = @lines_gffs[k].mrna_info(record.id)\n",
    "                tmp[:cds_count]   = mrna_stats.cds_count\n",
    "                tmp[:cds_max_gap] = mrna_stats.cds_max_gap\n",
    "                ret << tmp\n",
    "            end\n",
    "        end\n",
    "        ret\n",
    "    end\n",
    "    \n",
    "     attr_reader :lines_gffs\n",
    "    \n",
    "    def to_svg(mrna: \"Sm1_CDS.mrna1\", positions: false, out: nil)\n",
    "        p = Bio::Graphics::Page.new(width: 800,\n",
    "             height: 1000, \n",
    "              number_of_intervals:10,\n",
    "            background_color: \"white\"\n",
    "            )\n",
    "        each_gff do |k,v|\n",
    "            generic_track = p.add_track(:glyph => :generic, \n",
    "                            :name => k, \n",
    "                            :label => true  )\n",
    "            v.cds_to_print(mrna).each do |cds|\n",
    "                \n",
    "                f_id = positions ? cds.offset_start : nil\n",
    "                feature = Bio::Graphics::MiniFeature.new(start: cds.start, \n",
    "                    end: cds.end,\n",
    "                    fill_color: cds.color, \n",
    "                    id: f_id)\n",
    "                generic_track.add(feature) \n",
    "            end \n",
    "        end\n",
    "        \n",
    "         \n",
    "        p.write(out) if out\n",
    "        \n",
    "        p.get_markup\n",
    "    end\n",
    "end\n",
    "base_path=\"/Volumes/PanGenome/References\"\n",
    "lines = [\"cadenza\",\"claire\", \"paragon\",\"robigus\",\"weebil\"]\n",
    "gffs = MultipleGFFs.new(folder: \"#{base_path}/releaseV3/gff\",\n",
    "    lines:lines, \n",
    "    suffix:\".gff.gz\", \n",
    "    is_gz:true )\n",
    "puts \"\"\n",
    "\n",
    "cad = gffs[\"cadenza\"]\n",
    "\n",
    "valid_genes = Set.new\n",
    "gffs.each_gff do |k,v|\n",
    "    v.each_gene do |g|\n",
    "        valid_genes << g.id\n",
    "    end\n",
    "end\n",
    "\n",
    "valid_genes.size"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This file contains the source genes for the EI assemblies. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "638111"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ei_mapping = Daru::DataFrame.from_csv(\"#{base_path}/releaseV3/gff/UK.geneid2CSinformant.txt\", col_sep: \"\\t\")\n",
    "ei_mapping.size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<b> Daru::DataFrame(10x7) </b>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "  \n",
       "    <tr>\n",
       "      <th></th>\n",
       "      \n",
       "        <th>line</th>\n",
       "      \n",
       "        <th>projected_gene</th>\n",
       "      \n",
       "        <th>transcript</th>\n",
       "      \n",
       "        <th>chromosome</th>\n",
       "      \n",
       "        <th>gene</th>\n",
       "      \n",
       "        <th>scaffold</th>\n",
       "      \n",
       "        <th>is_gene</th>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "</thead>\n",
       "  <tbody>\n",
       "  \n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      \n",
       "        <td>cadenza</td>\n",
       "      \n",
       "        <td>TraesCAD_scaffold_000001_01G000100</td>\n",
       "      \n",
       "        <td>TraesCS1B02G148300.1</td>\n",
       "      \n",
       "        <td>1B</td>\n",
       "      \n",
       "        <td>TraesCS1B02G148300</td>\n",
       "      \n",
       "        <td>cad_scaffold_000001</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      \n",
       "        <td>cadenza</td>\n",
       "      \n",
       "        <td>TraesCAD_scaffold_000001_01G000200</td>\n",
       "      \n",
       "        <td>TraesCSU02G134200.1</td>\n",
       "      \n",
       "        <td>U</td>\n",
       "      \n",
       "        <td>TraesCSU02G134200</td>\n",
       "      \n",
       "        <td>cad_scaffold_000001</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      \n",
       "        <td>cadenza</td>\n",
       "      \n",
       "        <td>TraesCAD_scaffold_000001_01G000300</td>\n",
       "      \n",
       "        <td>TraesCS7B02G186200.1</td>\n",
       "      \n",
       "        <td>7B</td>\n",
       "      \n",
       "        <td>TraesCS7B02G186200</td>\n",
       "      \n",
       "        <td>cad_scaffold_000001</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      \n",
       "        <td>cadenza</td>\n",
       "      \n",
       "        <td>TraesCAD_scaffold_000003_01G000100</td>\n",
       "      \n",
       "        <td>TraesCS6D02G266000.1</td>\n",
       "      \n",
       "        <td>6D</td>\n",
       "      \n",
       "        <td>TraesCS6D02G266000</td>\n",
       "      \n",
       "        <td>cad_scaffold_000003</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      \n",
       "        <td>cadenza</td>\n",
       "      \n",
       "        <td>TraesCAD_scaffold_000003_01G000200</td>\n",
       "      \n",
       "        <td>TraesCS6B02G208700.1</td>\n",
       "      \n",
       "        <td>6B</td>\n",
       "      \n",
       "        <td>TraesCS6B02G208700</td>\n",
       "      \n",
       "        <td>cad_scaffold_000003</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      \n",
       "        <td>cadenza</td>\n",
       "      \n",
       "        <td>TraesCAD_scaffold_000003_01G000300</td>\n",
       "      \n",
       "        <td>TraesCS6B02G208800.1</td>\n",
       "      \n",
       "        <td>6B</td>\n",
       "      \n",
       "        <td>TraesCS6B02G208800</td>\n",
       "      \n",
       "        <td>cad_scaffold_000003</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      \n",
       "        <td>cadenza</td>\n",
       "      \n",
       "        <td>TraesCAD_scaffold_000004_01G000100</td>\n",
       "      \n",
       "        <td>TraesCS7B02G122100.1</td>\n",
       "      \n",
       "        <td>7B</td>\n",
       "      \n",
       "        <td>TraesCS7B02G122100</td>\n",
       "      \n",
       "        <td>cad_scaffold_000004</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      \n",
       "        <td>cadenza</td>\n",
       "      \n",
       "        <td>TraesCAD_scaffold_000004_01G000200</td>\n",
       "      \n",
       "        <td>TraesCS6D02G188400.1</td>\n",
       "      \n",
       "        <td>6D</td>\n",
       "      \n",
       "        <td>TraesCS6D02G188400</td>\n",
       "      \n",
       "        <td>cad_scaffold_000004</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      \n",
       "        <td>cadenza</td>\n",
       "      \n",
       "        <td>TraesCAD_scaffold_000004_01G000300</td>\n",
       "      \n",
       "        <td>TraesCS5D02G258700.1</td>\n",
       "      \n",
       "        <td>5D</td>\n",
       "      \n",
       "        <td>TraesCS5D02G258700</td>\n",
       "      \n",
       "        <td>cad_scaffold_000004</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      \n",
       "        <td>cadenza</td>\n",
       "      \n",
       "        <td>TraesCAD_scaffold_000004_01G000400</td>\n",
       "      \n",
       "        <td>TraesCS6B02G357700.1</td>\n",
       "      \n",
       "        <td>6B</td>\n",
       "      \n",
       "        <td>TraesCS6B02G357700</td>\n",
       "      \n",
       "        <td>cad_scaffold_000004</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "\n",
       "  \n",
       "</tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "#<Daru::DataFrame(10x7)>\n",
       "                  line projected_ transcript chromosome       gene   scaffold    is_gene\n",
       "          0    cadenza TraesCAD_s TraesCS1B0         1B TraesCS1B0 cad_scaffo       true\n",
       "          1    cadenza TraesCAD_s TraesCSU02          U TraesCSU02 cad_scaffo       true\n",
       "          2    cadenza TraesCAD_s TraesCS7B0         7B TraesCS7B0 cad_scaffo       true\n",
       "          3    cadenza TraesCAD_s TraesCS6D0         6D TraesCS6D0 cad_scaffo       true\n",
       "          4    cadenza TraesCAD_s TraesCS6B0         6B TraesCS6B0 cad_scaffo       true\n",
       "          5    cadenza TraesCAD_s TraesCS6B0         6B TraesCS6B0 cad_scaffo       true\n",
       "          6    cadenza TraesCAD_s TraesCS7B0         7B TraesCS7B0 cad_scaffo       true\n",
       "          7    cadenza TraesCAD_s TraesCS6D0         6D TraesCS6D0 cad_scaffo       true\n",
       "          8    cadenza TraesCAD_s TraesCS5D0         5D TraesCS5D0 cad_scaffo       true\n",
       "          9    cadenza TraesCAD_s TraesCS6B0         6B TraesCS6B0 cad_scaffo       true"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ei_mapping[\"chromosome\"] = ei_mapping.map(:row) do |row|\n",
    "  parseTranscript(row[\"transcript\"]).chromosome\n",
    "end\n",
    "\n",
    "ei_mapping[\"gene\"] = ei_mapping.map(:row) do |row|\n",
    "  parseTranscript(row[\"transcript\"]).gene\n",
    "end\n",
    "ei_mapping[\"scaffold\"] =  ei_mapping.map(:row) do |row|\n",
    "    parseEITranscript(row[\"projected_gene\"]).chromosome\n",
    "end\n",
    "ei_mapping[\"is_gene\"] =  ei_mapping.map(:row) do |row|\n",
    "    valid_genes.include? row[\"projected_gene\"]\n",
    "end\n",
    "ei_mapping.first 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<b> Daru::DataFrame(1x7) </b>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "  \n",
       "    <tr>\n",
       "      <th></th>\n",
       "      \n",
       "        <th>line</th>\n",
       "      \n",
       "        <th>projected_gene</th>\n",
       "      \n",
       "        <th>transcript</th>\n",
       "      \n",
       "        <th>chromosome</th>\n",
       "      \n",
       "        <th>gene</th>\n",
       "      \n",
       "        <th>scaffold</th>\n",
       "      \n",
       "        <th>is_gene</th>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "</thead>\n",
       "  <tbody>\n",
       "  \n",
       "    <tr>\n",
       "      <td>510301</td>\n",
       "      \n",
       "        <td>weebil</td>\n",
       "      \n",
       "        <td>TraesWEE_scaffold_000001_01G000100</td>\n",
       "      \n",
       "        <td>TraesCS6D02G266000.1</td>\n",
       "      \n",
       "        <td>6D</td>\n",
       "      \n",
       "        <td>TraesCS6D02G266000</td>\n",
       "      \n",
       "        <td>wee_scaffold_000001</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "\n",
       "  \n",
       "</tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "#<Daru::DataFrame(1x7)>\n",
       "                  line projected_ transcript chromosome       gene   scaffold    is_gene\n",
       "     510301     weebil TraesWEE_s TraesCS6D0         6D TraesCS6D0 wee_scaffo       true"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ei_mapping.filter(:row) do |row|\n",
    "  [\"TraesWEE_scaffold_000001_01G000100\"].include? row[\"projected_gene\"] \n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "638111"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ei_mapping_sorted = ei_mapping.sort([\"scaffold\"])\n",
    "ei_mapping_sorted.size"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following section takes the sorted table by scaffold to get the chromosome consensus of each scaffold. It is based on the majority of the chromosome of the source gene "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<b> Daru::DataFrame(10x5) </b>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "  \n",
       "    <tr>\n",
       "      <th></th>\n",
       "      \n",
       "        <th>scaffold</th>\n",
       "      \n",
       "        <th>chr</th>\n",
       "      \n",
       "        <th>hits</th>\n",
       "      \n",
       "        <th>total_genes</th>\n",
       "      \n",
       "        <th>percentage</th>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "</thead>\n",
       "  <tbody>\n",
       "  \n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      \n",
       "        <td>cad_scaffold_000001</td>\n",
       "      \n",
       "        <td>1B</td>\n",
       "      \n",
       "        <td>1</td>\n",
       "      \n",
       "        <td>3</td>\n",
       "      \n",
       "        <td>33</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      \n",
       "        <td>cad_scaffold_000003</td>\n",
       "      \n",
       "        <td>6B</td>\n",
       "      \n",
       "        <td>2</td>\n",
       "      \n",
       "        <td>3</td>\n",
       "      \n",
       "        <td>66</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      \n",
       "        <td>cad_scaffold_000004</td>\n",
       "      \n",
       "        <td>6B</td>\n",
       "      \n",
       "        <td>2</td>\n",
       "      \n",
       "        <td>5</td>\n",
       "      \n",
       "        <td>40</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      \n",
       "        <td>cad_scaffold_000005</td>\n",
       "      \n",
       "        <td>7A</td>\n",
       "      \n",
       "        <td>1</td>\n",
       "      \n",
       "        <td>2</td>\n",
       "      \n",
       "        <td>50</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      \n",
       "        <td>cad_scaffold_000006</td>\n",
       "      \n",
       "        <td>1B</td>\n",
       "      \n",
       "        <td>1</td>\n",
       "      \n",
       "        <td>2</td>\n",
       "      \n",
       "        <td>50</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      \n",
       "        <td>cad_scaffold_000007</td>\n",
       "      \n",
       "        <td>2B</td>\n",
       "      \n",
       "        <td>4</td>\n",
       "      \n",
       "        <td>6</td>\n",
       "      \n",
       "        <td>66</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      \n",
       "        <td>cad_scaffold_000008</td>\n",
       "      \n",
       "        <td>7B</td>\n",
       "      \n",
       "        <td>5</td>\n",
       "      \n",
       "        <td>5</td>\n",
       "      \n",
       "        <td>100</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      \n",
       "        <td>cad_scaffold_000009</td>\n",
       "      \n",
       "        <td>2B</td>\n",
       "      \n",
       "        <td>9</td>\n",
       "      \n",
       "        <td>10</td>\n",
       "      \n",
       "        <td>90</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      \n",
       "        <td>cad_scaffold_000010</td>\n",
       "      \n",
       "        <td>1B</td>\n",
       "      \n",
       "        <td>2</td>\n",
       "      \n",
       "        <td>2</td>\n",
       "      \n",
       "        <td>100</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      \n",
       "        <td>cad_scaffold_000011</td>\n",
       "      \n",
       "        <td>7B</td>\n",
       "      \n",
       "        <td>1</td>\n",
       "      \n",
       "        <td>1</td>\n",
       "      \n",
       "        <td>100</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "\n",
       "  \n",
       "</tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "#<Daru::DataFrame(10x5)>\n",
       "              scaffold        chr       hits total_gene percentage\n",
       "          0 cad_scaffo         1B          1          3         33\n",
       "          1 cad_scaffo         6B          2          3         66\n",
       "          2 cad_scaffo         6B          2          5         40\n",
       "          3 cad_scaffo         7A          1          2         50\n",
       "          4 cad_scaffo         1B          1          2         50\n",
       "          5 cad_scaffo         2B          4          6         66\n",
       "          6 cad_scaffo         7B          5          5        100\n",
       "          7 cad_scaffo         2B          9         10         90\n",
       "          8 cad_scaffo         1B          2          2        100\n",
       "          9 cad_scaffo         7B          1          1        100"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "i = 0\n",
    "current_scaffold = \"\"\n",
    "current_chromosomes = Hash.new {|h,k| h[k] = 0}\n",
    "consensus_chromosome = {}\n",
    "ei_mapping_sorted.each_row do |row|\n",
    "    if current_scaffold != row[\"scaffold\"]\n",
    "        if current_chromosomes.size > 0\n",
    "            max_v = current_chromosomes.max_by{|k,v| v}\n",
    "            consensus_chromosome[current_scaffold] = [\n",
    "              current_scaffold, \n",
    "              max_v[0], \n",
    "              max_v[1],  \n",
    "              current_chromosomes.values.reduce(:+),\n",
    "              100 * max_v[1] / current_chromosomes.values.reduce(:+)\n",
    "            ]\n",
    "        end\n",
    "        current_scaffold  = row[\"scaffold\"] \n",
    "        current_chromosomes = Hash.new {|h,k| h[k] = 0}\n",
    "        i+=1\n",
    "        \n",
    "    end \n",
    "    current_chromosomes[row[\"chromosome\"]] += 1 if row[\"is_gene\"]\n",
    "    \n",
    "end\n",
    "max_v = current_chromosomes.max_by{|k,v| v}\n",
    "\n",
    "consensus_chromosome[current_scaffold] = [\n",
    "  current_scaffold, \n",
    "  max_v[0],\n",
    "  max_v[1],\n",
    "  current_chromosomes.values.reduce(:+), \n",
    "  100 * max_v[1] / current_chromosomes.values.reduce(:+)\n",
    "]\n",
    "\n",
    "consensus_chr_df = Daru::DataFrame.rows(consensus_chromosome.values , \n",
    "    order: [:scaffold, :chr, :hits, :total_genes, :percentage])\n",
    "consensus_chr_df.head"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{30=>4156, 65=>12759, 40=>895, 50=>28408, 100=>256295, 90=>313, 80=>4506, 45=>14, 75=>6094, 60=>1879, 25=>731, 70=>624, 55=>398, 85=>1420, 35=>69, 95=>4, 20=>108, 15=>14, 10=>1}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "consensus_genes_per_scaff = Hash.new {|h,k| h[k] = 0}\n",
    "genes_per_scaff = Hash.new {|h,k| h[k] = 0}\n",
    "percentage_cons_per_scaff = Hash.new {|h,k| h[k] = 0}\n",
    "\n",
    "consensus_chr_df.each_row do |row|\n",
    "    consensus_genes_per_scaff[  row[:hits]] += 1\n",
    "    genes_per_scaff[ row[:total_genes]] += 1\n",
    "    percentage_cons_per_scaff[ 5 * (row[:percentage] / 5 )] += 1\n",
    "end\n",
    "\n",
    "percentage_cons_per_scaff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div id='vis-2df2e404-24d3-4e4a-acfb-739db8466683'></div>\n",
       "<script>\n",
       "(function(){\n",
       "    var render = function(){\n",
       "        var model = {\"panes\":[{\"diagrams\":[{\"type\":\"line\",\"options\":{\"x\":\"data0\",\"y\":\"data1\"},\"data\":\"1c3e99c5-88a4-4c96-afcc-341ab367dd9e\"}],\"options\":{\"x_label\":\"Percentage of genes corresponding to consensus\",\"y_label\":\"Count\",\"zoom\":true,\"width\":700,\"xrange\":[10,100],\"yrange\":[1,256295]}}],\"data\":{\"1c3e99c5-88a4-4c96-afcc-341ab367dd9e\":[{\"data0\":10,\"data1\":1},{\"data0\":15,\"data1\":14},{\"data0\":20,\"data1\":108},{\"data0\":25,\"data1\":731},{\"data0\":30,\"data1\":4156},{\"data0\":35,\"data1\":69},{\"data0\":40,\"data1\":895},{\"data0\":45,\"data1\":14},{\"data0\":50,\"data1\":28408},{\"data0\":55,\"data1\":398},{\"data0\":60,\"data1\":1879},{\"data0\":65,\"data1\":12759},{\"data0\":70,\"data1\":624},{\"data0\":75,\"data1\":6094},{\"data0\":80,\"data1\":4506},{\"data0\":85,\"data1\":1420},{\"data0\":90,\"data1\":313},{\"data0\":95,\"data1\":4},{\"data0\":100,\"data1\":256295}]},\"extension\":[]}\n",
       "        var id_name = '#vis-2df2e404-24d3-4e4a-acfb-739db8466683';\n",
       "        Nyaplot.core.parse(model, id_name);\n",
       "\n",
       "        require(['downloadable'], function(downloadable){\n",
       "          var svg = d3.select(id_name).select(\"svg\");\n",
       "\t  if(!svg.empty())\n",
       "\t    svg.call(downloadable().filename('fig'));\n",
       "\t});\n",
       "    };\n",
       "    if(window['Nyaplot']==undefined){\n",
       "        window.addEventListener('load_nyaplot', render, false);\n",
       "\treturn;\n",
       "    } else {\n",
       "       render();\n",
       "    }\n",
       "})();\n",
       "</script>\n"
      ],
      "text/plain": [
       "#<Nyaplot::Frame:0x00007fce76335c60 @properties={:panes=>[#<Nyaplot::Plot:0x00007fce7632c408 @properties={:diagrams=>[#<Nyaplot::Diagram:0x00007fce76336318 @properties={:type=>:line, :options=>{:x=>\"data0\", :y=>\"data1\"}, :data=>\"1c3e99c5-88a4-4c96-afcc-341ab367dd9e\"}, @xrange=[10, 100], @yrange=[1, 256295]>], :options=>{:x_label=>\"Percentage of genes corresponding to consensus\", :y_label=>\"Count\", :zoom=>true, :width=>700, :xrange=>[10, 100], :yrange=>[1, 256295]}}>], :data=>{\"1c3e99c5-88a4-4c96-afcc-341ab367dd9e\"=>#<Nyaplot::DataFrame:0x00007fce76337a88 @name=\"1c3e99c5-88a4-4c96-afcc-341ab367dd9e\", @rows=[{:data0=>10, :data1=>1}, {:data0=>15, :data1=>14}, {:data0=>20, :data1=>108}, {:data0=>25, :data1=>731}, {:data0=>30, :data1=>4156}, {:data0=>35, :data1=>69}, {:data0=>40, :data1=>895}, {:data0=>45, :data1=>14}, {:data0=>50, :data1=>28408}, {:data0=>55, :data1=>398}, {:data0=>60, :data1=>1879}, {:data0=>65, :data1=>12759}, {:data0=>70, :data1=>624}, {:data0=>75, :data1=>6094}, {:data0=>80, :data1=>4506}, {:data0=>85, :data1=>1420}, {:data0=>90, :data1=>313}, {:data0=>95, :data1=>4}, {:data0=>100, :data1=>256295}]>}, :extension=>[]}>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "percentage_cons_df = Daru::DataFrame.new({\n",
    "  percentage: percentage_cons_per_scaff.keys,\n",
    "  count: percentage_cons_per_scaff.values,\n",
    "  })\n",
    "percentage_cons_df.sort!([:percentage])\n",
    "plot = Nyaplot::Plot.new\n",
    "bar = plot.add(:line, percentage_cons_df[:percentage].to_a, percentage_cons_df[:count].to_a)\n",
    "plot.x_label \"Percentage of genes corresponding to consensus\"\n",
    "plot.y_label \"Count\"\n",
    "plot.show\n",
    "puts \"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "consensus_chr_df.write_csv(\"./EI_consensus_genes.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[\"cad_scaffold_000001\", \"1B\"]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cons_hash = {}\n",
    "consensus_chr_df.each_row{|row| cons_hash[row[:scaffold]] = row[:chr]}\n",
    "cons_hash.first\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Comparing each consensus chromosome to the gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<b> Daru::DataFrame(20x8) </b>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "  \n",
       "    <tr>\n",
       "      <th></th>\n",
       "      \n",
       "        <th>line</th>\n",
       "      \n",
       "        <th>projected_gene</th>\n",
       "      \n",
       "        <th>transcript</th>\n",
       "      \n",
       "        <th>chromosome</th>\n",
       "      \n",
       "        <th>gene</th>\n",
       "      \n",
       "        <th>scaffold</th>\n",
       "      \n",
       "        <th>is_gene</th>\n",
       "      \n",
       "        <th>is_chr_gene_consistent</th>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "</thead>\n",
       "  <tbody>\n",
       "  \n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      \n",
       "        <td>cadenza</td>\n",
       "      \n",
       "        <td>TraesCAD_scaffold_000001_01G000100</td>\n",
       "      \n",
       "        <td>TraesCS1B02G148300.1</td>\n",
       "      \n",
       "        <td>1B</td>\n",
       "      \n",
       "        <td>TraesCS1B02G148300</td>\n",
       "      \n",
       "        <td>cad_scaffold_000001</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      \n",
       "        <td>cadenza</td>\n",
       "      \n",
       "        <td>TraesCAD_scaffold_000001_01G000200</td>\n",
       "      \n",
       "        <td>TraesCSU02G134200.1</td>\n",
       "      \n",
       "        <td>U</td>\n",
       "      \n",
       "        <td>TraesCSU02G134200</td>\n",
       "      \n",
       "        <td>cad_scaffold_000001</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>false</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      \n",
       "        <td>cadenza</td>\n",
       "      \n",
       "        <td>TraesCAD_scaffold_000001_01G000300</td>\n",
       "      \n",
       "        <td>TraesCS7B02G186200.1</td>\n",
       "      \n",
       "        <td>7B</td>\n",
       "      \n",
       "        <td>TraesCS7B02G186200</td>\n",
       "      \n",
       "        <td>cad_scaffold_000001</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>false</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      \n",
       "        <td>cadenza</td>\n",
       "      \n",
       "        <td>TraesCAD_scaffold_000003_01G000100</td>\n",
       "      \n",
       "        <td>TraesCS6D02G266000.1</td>\n",
       "      \n",
       "        <td>6D</td>\n",
       "      \n",
       "        <td>TraesCS6D02G266000</td>\n",
       "      \n",
       "        <td>cad_scaffold_000003</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>false</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      \n",
       "        <td>cadenza</td>\n",
       "      \n",
       "        <td>TraesCAD_scaffold_000003_01G000200</td>\n",
       "      \n",
       "        <td>TraesCS6B02G208700.1</td>\n",
       "      \n",
       "        <td>6B</td>\n",
       "      \n",
       "        <td>TraesCS6B02G208700</td>\n",
       "      \n",
       "        <td>cad_scaffold_000003</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      \n",
       "        <td>cadenza</td>\n",
       "      \n",
       "        <td>TraesCAD_scaffold_000003_01G000300</td>\n",
       "      \n",
       "        <td>TraesCS6B02G208800.1</td>\n",
       "      \n",
       "        <td>6B</td>\n",
       "      \n",
       "        <td>TraesCS6B02G208800</td>\n",
       "      \n",
       "        <td>cad_scaffold_000003</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      \n",
       "        <td>cadenza</td>\n",
       "      \n",
       "        <td>TraesCAD_scaffold_000004_01G000100</td>\n",
       "      \n",
       "        <td>TraesCS7B02G122100.1</td>\n",
       "      \n",
       "        <td>7B</td>\n",
       "      \n",
       "        <td>TraesCS7B02G122100</td>\n",
       "      \n",
       "        <td>cad_scaffold_000004</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>false</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      \n",
       "        <td>cadenza</td>\n",
       "      \n",
       "        <td>TraesCAD_scaffold_000004_01G000200</td>\n",
       "      \n",
       "        <td>TraesCS6D02G188400.1</td>\n",
       "      \n",
       "        <td>6D</td>\n",
       "      \n",
       "        <td>TraesCS6D02G188400</td>\n",
       "      \n",
       "        <td>cad_scaffold_000004</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>false</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      \n",
       "        <td>cadenza</td>\n",
       "      \n",
       "        <td>TraesCAD_scaffold_000004_01G000300</td>\n",
       "      \n",
       "        <td>TraesCS5D02G258700.1</td>\n",
       "      \n",
       "        <td>5D</td>\n",
       "      \n",
       "        <td>TraesCS5D02G258700</td>\n",
       "      \n",
       "        <td>cad_scaffold_000004</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>false</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      \n",
       "        <td>cadenza</td>\n",
       "      \n",
       "        <td>TraesCAD_scaffold_000004_01G000400</td>\n",
       "      \n",
       "        <td>TraesCS6B02G357700.1</td>\n",
       "      \n",
       "        <td>6B</td>\n",
       "      \n",
       "        <td>TraesCS6B02G357700</td>\n",
       "      \n",
       "        <td>cad_scaffold_000004</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>10</td>\n",
       "      \n",
       "        <td>cadenza</td>\n",
       "      \n",
       "        <td>TraesCAD_scaffold_000004_01G000500</td>\n",
       "      \n",
       "        <td>TraesCS6B02G357800.1</td>\n",
       "      \n",
       "        <td>6B</td>\n",
       "      \n",
       "        <td>TraesCS6B02G357800</td>\n",
       "      \n",
       "        <td>cad_scaffold_000004</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>11</td>\n",
       "      \n",
       "        <td>cadenza</td>\n",
       "      \n",
       "        <td>TraesCAD_scaffold_000005_01G000100</td>\n",
       "      \n",
       "        <td>TraesCS7A02G427500.1</td>\n",
       "      \n",
       "        <td>7A</td>\n",
       "      \n",
       "        <td>TraesCS7A02G427500</td>\n",
       "      \n",
       "        <td>cad_scaffold_000005</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>12</td>\n",
       "      \n",
       "        <td>cadenza</td>\n",
       "      \n",
       "        <td>TraesCAD_scaffold_000005_01G000200</td>\n",
       "      \n",
       "        <td>TraesCS4A02G150700.1</td>\n",
       "      \n",
       "        <td>4A</td>\n",
       "      \n",
       "        <td>TraesCS4A02G150700</td>\n",
       "      \n",
       "        <td>cad_scaffold_000005</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>false</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>13</td>\n",
       "      \n",
       "        <td>cadenza</td>\n",
       "      \n",
       "        <td>TraesCAD_scaffold_000006_01G000100</td>\n",
       "      \n",
       "        <td>TraesCS1B02G148400.1</td>\n",
       "      \n",
       "        <td>1B</td>\n",
       "      \n",
       "        <td>TraesCS1B02G148400</td>\n",
       "      \n",
       "        <td>cad_scaffold_000006</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>14</td>\n",
       "      \n",
       "        <td>cadenza</td>\n",
       "      \n",
       "        <td>TraesCAD_scaffold_000006_01G000200</td>\n",
       "      \n",
       "        <td>TraesCS6B02G063600.1</td>\n",
       "      \n",
       "        <td>6B</td>\n",
       "      \n",
       "        <td>TraesCS6B02G063600</td>\n",
       "      \n",
       "        <td>cad_scaffold_000006</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>false</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>15</td>\n",
       "      \n",
       "        <td>cadenza</td>\n",
       "      \n",
       "        <td>TraesCAD_scaffold_000007_01G000100</td>\n",
       "      \n",
       "        <td>TraesCS4B02G081300.1</td>\n",
       "      \n",
       "        <td>4B</td>\n",
       "      \n",
       "        <td>TraesCS4B02G081300</td>\n",
       "      \n",
       "        <td>cad_scaffold_000007</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>false</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>16</td>\n",
       "      \n",
       "        <td>cadenza</td>\n",
       "      \n",
       "        <td>TraesCAD_scaffold_000007_01G000200</td>\n",
       "      \n",
       "        <td>TraesCS2B02G399100.1</td>\n",
       "      \n",
       "        <td>2B</td>\n",
       "      \n",
       "        <td>TraesCS2B02G399100</td>\n",
       "      \n",
       "        <td>cad_scaffold_000007</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>17</td>\n",
       "      \n",
       "        <td>cadenza</td>\n",
       "      \n",
       "        <td>TraesCAD_scaffold_000007_01G000300</td>\n",
       "      \n",
       "        <td>TraesCS2B02G399000.1</td>\n",
       "      \n",
       "        <td>2B</td>\n",
       "      \n",
       "        <td>TraesCS2B02G399000</td>\n",
       "      \n",
       "        <td>cad_scaffold_000007</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>18</td>\n",
       "      \n",
       "        <td>cadenza</td>\n",
       "      \n",
       "        <td>TraesCAD_scaffold_000007_01G000400</td>\n",
       "      \n",
       "        <td>TraesCS2B02G398900.1</td>\n",
       "      \n",
       "        <td>2B</td>\n",
       "      \n",
       "        <td>TraesCS2B02G398900</td>\n",
       "      \n",
       "        <td>cad_scaffold_000007</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>19</td>\n",
       "      \n",
       "        <td>cadenza</td>\n",
       "      \n",
       "        <td>TraesCAD_scaffold_000007_01G000500</td>\n",
       "      \n",
       "        <td>TraesCS2D02G466900.1</td>\n",
       "      \n",
       "        <td>2D</td>\n",
       "      \n",
       "        <td>TraesCS2D02G466900</td>\n",
       "      \n",
       "        <td>cad_scaffold_000007</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>false</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "\n",
       "  \n",
       "</tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "#<Daru::DataFrame(20x8)>\n",
       "                  line projected_ transcript chromosome       gene   scaffold    is_gene is_chr_gen\n",
       "          0    cadenza TraesCAD_s TraesCS1B0         1B TraesCS1B0 cad_scaffo       true       true\n",
       "          1    cadenza TraesCAD_s TraesCSU02          U TraesCSU02 cad_scaffo       true      false\n",
       "          2    cadenza TraesCAD_s TraesCS7B0         7B TraesCS7B0 cad_scaffo       true      false\n",
       "          3    cadenza TraesCAD_s TraesCS6D0         6D TraesCS6D0 cad_scaffo       true      false\n",
       "          4    cadenza TraesCAD_s TraesCS6B0         6B TraesCS6B0 cad_scaffo       true       true\n",
       "          5    cadenza TraesCAD_s TraesCS6B0         6B TraesCS6B0 cad_scaffo       true       true\n",
       "          6    cadenza TraesCAD_s TraesCS7B0         7B TraesCS7B0 cad_scaffo       true      false\n",
       "          7    cadenza TraesCAD_s TraesCS6D0         6D TraesCS6D0 cad_scaffo       true      false\n",
       "          8    cadenza TraesCAD_s TraesCS5D0         5D TraesCS5D0 cad_scaffo       true      false\n",
       "          9    cadenza TraesCAD_s TraesCS6B0         6B TraesCS6B0 cad_scaffo       true       true\n",
       "         10    cadenza TraesCAD_s TraesCS6B0         6B TraesCS6B0 cad_scaffo       true       true\n",
       "         11    cadenza TraesCAD_s TraesCS7A0         7A TraesCS7A0 cad_scaffo       true       true\n",
       "         12    cadenza TraesCAD_s TraesCS4A0         4A TraesCS4A0 cad_scaffo       true      false\n",
       "         13    cadenza TraesCAD_s TraesCS1B0         1B TraesCS1B0 cad_scaffo       true       true\n",
       "         14    cadenza TraesCAD_s TraesCS6B0         6B TraesCS6B0 cad_scaffo       true      false\n",
       "        ...        ...        ...        ...        ...        ...        ...        ...        ..."
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ei_mapping[\"is_chr_gene_consistent\"] =  ei_mapping.map(:row) do |row|\n",
    "    cons_hash[row[\"scaffold\"]] == row[\"chromosome\"]\n",
    "end\n",
    "ei_mapping.first 20"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Comparing that it is not a pseudogene and that it is consistent. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "455268\n"
     ]
    }
   ],
   "source": [
    "gene_count = Hash.new{|h,k| h[k] = 0}\n",
    "ei_mapping.each_row do |row|\n",
    "    next unless row[\"is_gene\"] and row[\"is_chr_gene_consistent\"]\n",
    "    gene_count[row[\"line\"] + \"_\" + row[\"gene\"]] += 1\n",
    "end\n",
    "puts gene_count.size"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, comparing that the gene is mapped only once (Removes posible dupplications or genes that are ambigous)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<b> Daru::DataFrame(10x9) </b>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "  \n",
       "    <tr>\n",
       "      <th></th>\n",
       "      \n",
       "        <th>line</th>\n",
       "      \n",
       "        <th>projected_gene</th>\n",
       "      \n",
       "        <th>transcript</th>\n",
       "      \n",
       "        <th>chromosome</th>\n",
       "      \n",
       "        <th>gene</th>\n",
       "      \n",
       "        <th>scaffold</th>\n",
       "      \n",
       "        <th>is_gene</th>\n",
       "      \n",
       "        <th>is_chr_gene_consistent</th>\n",
       "      \n",
       "        <th>is_unique</th>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "</thead>\n",
       "  <tbody>\n",
       "  \n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      \n",
       "        <td>cadenza</td>\n",
       "      \n",
       "        <td>TraesCAD_scaffold_000001_01G000100</td>\n",
       "      \n",
       "        <td>TraesCS1B02G148300.1</td>\n",
       "      \n",
       "        <td>1B</td>\n",
       "      \n",
       "        <td>TraesCS1B02G148300</td>\n",
       "      \n",
       "        <td>cad_scaffold_000001</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>false</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      \n",
       "        <td>cadenza</td>\n",
       "      \n",
       "        <td>TraesCAD_scaffold_000001_01G000200</td>\n",
       "      \n",
       "        <td>TraesCSU02G134200.1</td>\n",
       "      \n",
       "        <td>U</td>\n",
       "      \n",
       "        <td>TraesCSU02G134200</td>\n",
       "      \n",
       "        <td>cad_scaffold_000001</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>false</td>\n",
       "      \n",
       "        <td>false</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      \n",
       "        <td>cadenza</td>\n",
       "      \n",
       "        <td>TraesCAD_scaffold_000001_01G000300</td>\n",
       "      \n",
       "        <td>TraesCS7B02G186200.1</td>\n",
       "      \n",
       "        <td>7B</td>\n",
       "      \n",
       "        <td>TraesCS7B02G186200</td>\n",
       "      \n",
       "        <td>cad_scaffold_000001</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>false</td>\n",
       "      \n",
       "        <td>false</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      \n",
       "        <td>cadenza</td>\n",
       "      \n",
       "        <td>TraesCAD_scaffold_000003_01G000100</td>\n",
       "      \n",
       "        <td>TraesCS6D02G266000.1</td>\n",
       "      \n",
       "        <td>6D</td>\n",
       "      \n",
       "        <td>TraesCS6D02G266000</td>\n",
       "      \n",
       "        <td>cad_scaffold_000003</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>false</td>\n",
       "      \n",
       "        <td>false</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      \n",
       "        <td>cadenza</td>\n",
       "      \n",
       "        <td>TraesCAD_scaffold_000003_01G000200</td>\n",
       "      \n",
       "        <td>TraesCS6B02G208700.1</td>\n",
       "      \n",
       "        <td>6B</td>\n",
       "      \n",
       "        <td>TraesCS6B02G208700</td>\n",
       "      \n",
       "        <td>cad_scaffold_000003</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>false</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      \n",
       "        <td>cadenza</td>\n",
       "      \n",
       "        <td>TraesCAD_scaffold_000003_01G000300</td>\n",
       "      \n",
       "        <td>TraesCS6B02G208800.1</td>\n",
       "      \n",
       "        <td>6B</td>\n",
       "      \n",
       "        <td>TraesCS6B02G208800</td>\n",
       "      \n",
       "        <td>cad_scaffold_000003</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      \n",
       "        <td>cadenza</td>\n",
       "      \n",
       "        <td>TraesCAD_scaffold_000004_01G000100</td>\n",
       "      \n",
       "        <td>TraesCS7B02G122100.1</td>\n",
       "      \n",
       "        <td>7B</td>\n",
       "      \n",
       "        <td>TraesCS7B02G122100</td>\n",
       "      \n",
       "        <td>cad_scaffold_000004</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>false</td>\n",
       "      \n",
       "        <td>false</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      \n",
       "        <td>cadenza</td>\n",
       "      \n",
       "        <td>TraesCAD_scaffold_000004_01G000200</td>\n",
       "      \n",
       "        <td>TraesCS6D02G188400.1</td>\n",
       "      \n",
       "        <td>6D</td>\n",
       "      \n",
       "        <td>TraesCS6D02G188400</td>\n",
       "      \n",
       "        <td>cad_scaffold_000004</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>false</td>\n",
       "      \n",
       "        <td>false</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      \n",
       "        <td>cadenza</td>\n",
       "      \n",
       "        <td>TraesCAD_scaffold_000004_01G000300</td>\n",
       "      \n",
       "        <td>TraesCS5D02G258700.1</td>\n",
       "      \n",
       "        <td>5D</td>\n",
       "      \n",
       "        <td>TraesCS5D02G258700</td>\n",
       "      \n",
       "        <td>cad_scaffold_000004</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>false</td>\n",
       "      \n",
       "        <td>false</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      \n",
       "        <td>cadenza</td>\n",
       "      \n",
       "        <td>TraesCAD_scaffold_000004_01G000400</td>\n",
       "      \n",
       "        <td>TraesCS6B02G357700.1</td>\n",
       "      \n",
       "        <td>6B</td>\n",
       "      \n",
       "        <td>TraesCS6B02G357700</td>\n",
       "      \n",
       "        <td>cad_scaffold_000004</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "\n",
       "  \n",
       "</tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "#<Daru::DataFrame(10x9)>\n",
       "                  line projected_ transcript chromosome       gene   scaffold    is_gene is_chr_gen  is_unique\n",
       "          0    cadenza TraesCAD_s TraesCS1B0         1B TraesCS1B0 cad_scaffo       true       true      false\n",
       "          1    cadenza TraesCAD_s TraesCSU02          U TraesCSU02 cad_scaffo       true      false      false\n",
       "          2    cadenza TraesCAD_s TraesCS7B0         7B TraesCS7B0 cad_scaffo       true      false      false\n",
       "          3    cadenza TraesCAD_s TraesCS6D0         6D TraesCS6D0 cad_scaffo       true      false      false\n",
       "          4    cadenza TraesCAD_s TraesCS6B0         6B TraesCS6B0 cad_scaffo       true       true      false\n",
       "          5    cadenza TraesCAD_s TraesCS6B0         6B TraesCS6B0 cad_scaffo       true       true       true\n",
       "          6    cadenza TraesCAD_s TraesCS7B0         7B TraesCS7B0 cad_scaffo       true      false      false\n",
       "          7    cadenza TraesCAD_s TraesCS6D0         6D TraesCS6D0 cad_scaffo       true      false      false\n",
       "          8    cadenza TraesCAD_s TraesCS5D0         5D TraesCS5D0 cad_scaffo       true      false      false\n",
       "          9    cadenza TraesCAD_s TraesCS6B0         6B TraesCS6B0 cad_scaffo       true       true       true"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ei_mapping[\"is_unique\"] =  ei_mapping.map(:row) do |row|\n",
    "    gene_count[row[\"line\"] + \"_\" + row[\"gene\"]] == 1 and  row[\"is_gene\"] and row[\"is_chr_gene_consistent\"] \n",
    "end\n",
    "ei_mapping.head"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "consensus_chr_df.write_csv(\"EI_consensus_genes.csv\")\n",
    "ei_mapping.write_csv(\"ei_mapping.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<b> Daru::Vector(5) </b>\n",
       "<table>\n",
       "  <thead>\n",
       "  \n",
       "    <tr>\n",
       "        <th> </th>\n",
       "        <th>line</th>\n",
       "    </tr>\n",
       "  \n",
       "</thead>\n",
       "  <tbody>\n",
       "  \n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>cadenza</td>\n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>128544</td>\n",
       "      <td>claire</td>\n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>255850</td>\n",
       "      <td>paragon</td>\n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>383153</td>\n",
       "      <td>robigus</td>\n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>510301</td>\n",
       "      <td>weebil</td>\n",
       "    </tr>\n",
       "  \n",
       "\n",
       "  \n",
       "</tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "#<Daru::Vector(5)>\n",
       "            line\n",
       "       0 cadenza\n",
       "  128544  claire\n",
       "  255850 paragon\n",
       "  383153 robigus\n",
       "  510301  weebil"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ei_mapping[\"line\"].uniq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PGSB Assemblies "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "lines = [\n",
    "    \"arinalrfor\",\n",
    "    \"chinese\",\n",
    "    \"jagger\",\n",
    "    \"julius\",\n",
    "    \"lancer\",\n",
    "    \"landmark\",\n",
    "    \"mace\",\n",
    "    \"norin61\",\n",
    "    \"spelta\",\n",
    "    \"stanley\",\n",
    "    \"sy_mattis\"\n",
    "    ]\n",
    "gffs_pgsb = MultipleGFFs.new(folder: \"#{base_path}/releasePGSBv2.0/gff\",\n",
    "    lines:lines, \n",
    "    suffix:\".gff.gz\", \n",
    "    is_gz:true )\n",
    "puts \"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<b> Daru::DataFrame(10x8) </b>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "  \n",
       "    <tr>\n",
       "      <th></th>\n",
       "      \n",
       "        <th>line</th>\n",
       "      \n",
       "        <th>projected_gene</th>\n",
       "      \n",
       "        <th>transcript</th>\n",
       "      \n",
       "        <th>chromosome</th>\n",
       "      \n",
       "        <th>gene</th>\n",
       "      \n",
       "        <th>scaffold</th>\n",
       "      \n",
       "        <th>is_gene</th>\n",
       "      \n",
       "        <th>is_chr_gene_consistent</th>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "</thead>\n",
       "  <tbody>\n",
       "  \n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      \n",
       "        <td>arinalrfor</td>\n",
       "      \n",
       "        <td>TraesARI1A01G000100</td>\n",
       "      \n",
       "        <td>TraesCS1A02G003546.1</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>TraesCS1A02G003546</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>false</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      \n",
       "        <td>arinalrfor</td>\n",
       "      \n",
       "        <td>TraesARI1A01G000200</td>\n",
       "      \n",
       "        <td>TraesCS1A02G004126.1</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>TraesCS1A02G004126</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      \n",
       "        <td>arinalrfor</td>\n",
       "      \n",
       "        <td>TraesARI1A01G000300</td>\n",
       "      \n",
       "        <td>TraesCS1A02G003600.1</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>TraesCS1A02G003600</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      \n",
       "        <td>arinalrfor</td>\n",
       "      \n",
       "        <td>TraesARI1A01G000400</td>\n",
       "      \n",
       "        <td>TraesCS1A02G003700.1</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>TraesCS1A02G003700</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      \n",
       "        <td>arinalrfor</td>\n",
       "      \n",
       "        <td>TraesARI1A01G000500</td>\n",
       "      \n",
       "        <td>TraesCS7A02G485500.1</td>\n",
       "      \n",
       "        <td>7A</td>\n",
       "      \n",
       "        <td>TraesCS7A02G485500</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>false</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      \n",
       "        <td>arinalrfor</td>\n",
       "      \n",
       "        <td>TraesARI1A01G000600</td>\n",
       "      \n",
       "        <td>TraesCS3A02G080300.1</td>\n",
       "      \n",
       "        <td>3A</td>\n",
       "      \n",
       "        <td>TraesCS3A02G080300</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>false</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      \n",
       "        <td>arinalrfor</td>\n",
       "      \n",
       "        <td>TraesARI1A01G000700</td>\n",
       "      \n",
       "        <td>TraesCS6D02G375000.1</td>\n",
       "      \n",
       "        <td>6D</td>\n",
       "      \n",
       "        <td>TraesCS6D02G375000</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>false</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      \n",
       "        <td>arinalrfor</td>\n",
       "      \n",
       "        <td>TraesARI1A01G000800</td>\n",
       "      \n",
       "        <td>TraesCSU02G097400.1</td>\n",
       "      \n",
       "        <td>U</td>\n",
       "      \n",
       "        <td>TraesCSU02G097400</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>false</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      \n",
       "        <td>arinalrfor</td>\n",
       "      \n",
       "        <td>TraesARI1A01G000900</td>\n",
       "      \n",
       "        <td>TraesCS1A02G019300.1</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>TraesCS1A02G019300</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      \n",
       "        <td>arinalrfor</td>\n",
       "      \n",
       "        <td>TraesARI1A01G001000</td>\n",
       "      \n",
       "        <td>TraesCS1A02G003500.1</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>TraesCS1A02G003500</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "\n",
       "  \n",
       "</tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "#<Daru::DataFrame(10x8)>\n",
       "                  line projected_ transcript chromosome       gene   scaffold    is_gene is_chr_gen\n",
       "          0 arinalrfor TraesARI1A TraesCS1A0         1A TraesCS1A0         1A      false       true\n",
       "          1 arinalrfor TraesARI1A TraesCS1A0         1A TraesCS1A0         1A       true       true\n",
       "          2 arinalrfor TraesARI1A TraesCS1A0         1A TraesCS1A0         1A       true       true\n",
       "          3 arinalrfor TraesARI1A TraesCS1A0         1A TraesCS1A0         1A       true       true\n",
       "          4 arinalrfor TraesARI1A TraesCS7A0         7A TraesCS7A0         1A       true      false\n",
       "          5 arinalrfor TraesARI1A TraesCS3A0         3A TraesCS3A0         1A       true      false\n",
       "          6 arinalrfor TraesARI1A TraesCS6D0         6D TraesCS6D0         1A       true      false\n",
       "          7 arinalrfor TraesARI1A TraesCSU02          U TraesCSU02         1A       true      false\n",
       "          8 arinalrfor TraesARI1A TraesCS1A0         1A TraesCS1A0         1A       true       true\n",
       "          9 arinalrfor TraesARI1A TraesCS1A0         1A TraesCS1A0         1A       true       true"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def parsePGSBTranscript name\n",
    "    arr=name.split(\".\")\n",
    "    match = /Traes(?<variety>[[:upper:]]{3})(?<chr>[[:alnum:]]{1,2})(?<ver>[[:digit:]]{2})G(?<count>[[:digit:]]+)(?<conf>[[:upper:]]*)/.match arr[0]\n",
    "    #match  = /Traes(?<variety>[[:upper:]]*)/.match arr[0]\n",
    "    #puts match.inspect\n",
    "    raise \"Unable to parse: #{name}\" unless match\n",
    "    Transcript.new(name, arr[0],match[:chr],match[:ver],match[:count],match[:variety], match[:conf],match[:count].to_i, arr[1])\n",
    "end\n",
    "rows = []\n",
    "gffs_pgsb.each_gff do |k,v|\n",
    "    v.each_gene do |record|\n",
    "        srcmodel = record.attributes[\"srcmodel\"]\n",
    "        parsed_proj = parsePGSBTranscript record.id\n",
    "        parsed_src = parseTranscript srcmodel\n",
    "        \n",
    "        rows << [k, \n",
    "            record.id, \n",
    "            record.attributes[\"srcmodel\"],\n",
    "            parsed_src.chromosome, \n",
    "            parsed_src.gene,\n",
    "            parsed_proj.chromosome,  \n",
    "            record.attributes[\"disruptedORF\"].nil?, \n",
    "            parsed_proj.chromosome == parsed_src.chromosome ]      \n",
    "    end\n",
    "end\n",
    "\n",
    "pgsb_df = Daru::DataFrame.rows(rows , \n",
    "    order: [\"line\", \"projected_gene\", \"transcript\", \"chromosome\", \"gene\", \"scaffold\", \"is_gene\", \"is_chr_gene_consistent\"])\n",
    "pgsb_df.head"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Validating that the chromosome from the source gene is consistent with the chromosome. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1069393\n"
     ]
    }
   ],
   "source": [
    "gene_count_pgsb = Hash.new{|h,k| h[k] = 0}\n",
    "pgsb_df.each_row do |row|\n",
    "    next unless row[\"is_gene\"] and row[\"is_chr_gene_consistent\"]\n",
    "    gene_count_pgsb[row[\"line\"] + \"_\" + row[\"gene\"]] += 1\n",
    "end\n",
    "puts gene_count_pgsb.size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<b> Daru::DataFrame(10x9) </b>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "  \n",
       "    <tr>\n",
       "      <th></th>\n",
       "      \n",
       "        <th>line</th>\n",
       "      \n",
       "        <th>projected_gene</th>\n",
       "      \n",
       "        <th>transcript</th>\n",
       "      \n",
       "        <th>chromosome</th>\n",
       "      \n",
       "        <th>gene</th>\n",
       "      \n",
       "        <th>scaffold</th>\n",
       "      \n",
       "        <th>is_gene</th>\n",
       "      \n",
       "        <th>is_chr_gene_consistent</th>\n",
       "      \n",
       "        <th>is_unique</th>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "</thead>\n",
       "  <tbody>\n",
       "  \n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      \n",
       "        <td>arinalrfor</td>\n",
       "      \n",
       "        <td>TraesARI1A01G000100</td>\n",
       "      \n",
       "        <td>TraesCS1A02G003546.1</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>TraesCS1A02G003546</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>false</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>false</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      \n",
       "        <td>arinalrfor</td>\n",
       "      \n",
       "        <td>TraesARI1A01G000200</td>\n",
       "      \n",
       "        <td>TraesCS1A02G004126.1</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>TraesCS1A02G004126</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      \n",
       "        <td>arinalrfor</td>\n",
       "      \n",
       "        <td>TraesARI1A01G000300</td>\n",
       "      \n",
       "        <td>TraesCS1A02G003600.1</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>TraesCS1A02G003600</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      \n",
       "        <td>arinalrfor</td>\n",
       "      \n",
       "        <td>TraesARI1A01G000400</td>\n",
       "      \n",
       "        <td>TraesCS1A02G003700.1</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>TraesCS1A02G003700</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      \n",
       "        <td>arinalrfor</td>\n",
       "      \n",
       "        <td>TraesARI1A01G000500</td>\n",
       "      \n",
       "        <td>TraesCS7A02G485500.1</td>\n",
       "      \n",
       "        <td>7A</td>\n",
       "      \n",
       "        <td>TraesCS7A02G485500</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>false</td>\n",
       "      \n",
       "        <td>false</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      \n",
       "        <td>arinalrfor</td>\n",
       "      \n",
       "        <td>TraesARI1A01G000600</td>\n",
       "      \n",
       "        <td>TraesCS3A02G080300.1</td>\n",
       "      \n",
       "        <td>3A</td>\n",
       "      \n",
       "        <td>TraesCS3A02G080300</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>false</td>\n",
       "      \n",
       "        <td>false</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      \n",
       "        <td>arinalrfor</td>\n",
       "      \n",
       "        <td>TraesARI1A01G000700</td>\n",
       "      \n",
       "        <td>TraesCS6D02G375000.1</td>\n",
       "      \n",
       "        <td>6D</td>\n",
       "      \n",
       "        <td>TraesCS6D02G375000</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>false</td>\n",
       "      \n",
       "        <td>false</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      \n",
       "        <td>arinalrfor</td>\n",
       "      \n",
       "        <td>TraesARI1A01G000800</td>\n",
       "      \n",
       "        <td>TraesCSU02G097400.1</td>\n",
       "      \n",
       "        <td>U</td>\n",
       "      \n",
       "        <td>TraesCSU02G097400</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>false</td>\n",
       "      \n",
       "        <td>false</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      \n",
       "        <td>arinalrfor</td>\n",
       "      \n",
       "        <td>TraesARI1A01G000900</td>\n",
       "      \n",
       "        <td>TraesCS1A02G019300.1</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>TraesCS1A02G019300</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      \n",
       "        <td>arinalrfor</td>\n",
       "      \n",
       "        <td>TraesARI1A01G001000</td>\n",
       "      \n",
       "        <td>TraesCS1A02G003500.1</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>TraesCS1A02G003500</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "\n",
       "  \n",
       "</tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "#<Daru::DataFrame(10x9)>\n",
       "                  line projected_ transcript chromosome       gene   scaffold    is_gene is_chr_gen  is_unique\n",
       "          0 arinalrfor TraesARI1A TraesCS1A0         1A TraesCS1A0         1A      false       true      false\n",
       "          1 arinalrfor TraesARI1A TraesCS1A0         1A TraesCS1A0         1A       true       true       true\n",
       "          2 arinalrfor TraesARI1A TraesCS1A0         1A TraesCS1A0         1A       true       true       true\n",
       "          3 arinalrfor TraesARI1A TraesCS1A0         1A TraesCS1A0         1A       true       true       true\n",
       "          4 arinalrfor TraesARI1A TraesCS7A0         7A TraesCS7A0         1A       true      false      false\n",
       "          5 arinalrfor TraesARI1A TraesCS3A0         3A TraesCS3A0         1A       true      false      false\n",
       "          6 arinalrfor TraesARI1A TraesCS6D0         6D TraesCS6D0         1A       true      false      false\n",
       "          7 arinalrfor TraesARI1A TraesCSU02          U TraesCSU02         1A       true      false      false\n",
       "          8 arinalrfor TraesARI1A TraesCS1A0         1A TraesCS1A0         1A       true       true       true\n",
       "          9 arinalrfor TraesARI1A TraesCS1A0         1A TraesCS1A0         1A       true       true       true"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pgsb_df[\"is_unique\"] =  pgsb_df.map(:row) do |row|\n",
    "    gene_count_pgsb[row[\"line\"] + \"_\" + row[\"gene\"]] == 1 and  row[\"is_gene\"] and row[\"is_chr_gene_consistent\"] \n",
    "end\n",
    "pgsb_df.head"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<b> Daru::DataFrame(10x9) </b>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "  \n",
       "    <tr>\n",
       "      <th></th>\n",
       "      \n",
       "        <th>line</th>\n",
       "      \n",
       "        <th>projected_gene</th>\n",
       "      \n",
       "        <th>transcript</th>\n",
       "      \n",
       "        <th>chromosome</th>\n",
       "      \n",
       "        <th>gene</th>\n",
       "      \n",
       "        <th>scaffold</th>\n",
       "      \n",
       "        <th>is_gene</th>\n",
       "      \n",
       "        <th>is_chr_gene_consistent</th>\n",
       "      \n",
       "        <th>is_unique</th>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "</thead>\n",
       "  <tbody>\n",
       "  \n",
       "    <tr>\n",
       "      <td>244046</td>\n",
       "      \n",
       "        <td>jagger</td>\n",
       "      \n",
       "        <td>TraesJAG1A01G000500</td>\n",
       "      \n",
       "        <td>TraesCS1A02G000100.1</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>TraesCS1A02G000100</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>363868</td>\n",
       "      \n",
       "        <td>julius</td>\n",
       "      \n",
       "        <td>TraesJUL1A01G036600</td>\n",
       "      \n",
       "        <td>TraesCS1A02G000100.1</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>TraesCS1A02G000100</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>482640</td>\n",
       "      \n",
       "        <td>lancer</td>\n",
       "      \n",
       "        <td>TraesLAC1A01G000300</td>\n",
       "      \n",
       "        <td>TraesCS1A02G000100.1</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>TraesCS1A02G000100</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>602686</td>\n",
       "      \n",
       "        <td>landmark</td>\n",
       "      \n",
       "        <td>TraesLDM1A01G000400</td>\n",
       "      \n",
       "        <td>TraesCS1A02G000100.1</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>TraesCS1A02G000100</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>721733</td>\n",
       "      \n",
       "        <td>mace</td>\n",
       "      \n",
       "        <td>TraesMAC1A01G002400</td>\n",
       "      \n",
       "        <td>TraesCS1A02G000100.1</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>TraesCS1A02G000100</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>841525</td>\n",
       "      \n",
       "        <td>norin61</td>\n",
       "      \n",
       "        <td>TraesNOR1A01G004400</td>\n",
       "      \n",
       "        <td>TraesCS1A02G000100.1</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>TraesCS1A02G000100</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>960218</td>\n",
       "      \n",
       "        <td>spelta</td>\n",
       "      \n",
       "        <td>TraesTSP1A01G000300</td>\n",
       "      \n",
       "        <td>TraesCS1A02G000100.1</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>TraesCS1A02G000100</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>1080407</td>\n",
       "      \n",
       "        <td>stanley</td>\n",
       "      \n",
       "        <td>TraesSTA1A01G008800</td>\n",
       "      \n",
       "        <td>TraesCS1A02G000100.1</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>TraesCS1A02G000100</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>1448033</td>\n",
       "      \n",
       "        <td>cadenza</td>\n",
       "      \n",
       "        <td>TraesCAD_scaffold_628837_01G000200</td>\n",
       "      \n",
       "        <td>TraesCS1A02G000200.1</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>TraesCS1A02G000200</td>\n",
       "      \n",
       "        <td>cad_scaffold_628837</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "    <tr>\n",
       "      <td>1569964</td>\n",
       "      \n",
       "        <td>claire</td>\n",
       "      \n",
       "        <td>TraesCLE_scaffold_182945_01G000100</td>\n",
       "      \n",
       "        <td>TraesCS1A02G000200.1</td>\n",
       "      \n",
       "        <td>1A</td>\n",
       "      \n",
       "        <td>TraesCS1A02G000200</td>\n",
       "      \n",
       "        <td>cle_scaffold_182945</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "        <td>true</td>\n",
       "      \n",
       "    </tr>\n",
       "  \n",
       "\n",
       "  \n",
       "</tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "#<Daru::DataFrame(10x9)>\n",
       "                  line projected_ transcript chromosome       gene   scaffold    is_gene is_chr_gen  is_unique\n",
       "     244046     jagger TraesJAG1A TraesCS1A0         1A TraesCS1A0         1A       true       true       true\n",
       "     363868     julius TraesJUL1A TraesCS1A0         1A TraesCS1A0         1A       true       true       true\n",
       "     482640     lancer TraesLAC1A TraesCS1A0         1A TraesCS1A0         1A       true       true       true\n",
       "     602686   landmark TraesLDM1A TraesCS1A0         1A TraesCS1A0         1A       true       true       true\n",
       "     721733       mace TraesMAC1A TraesCS1A0         1A TraesCS1A0         1A       true       true       true\n",
       "     841525    norin61 TraesNOR1A TraesCS1A0         1A TraesCS1A0         1A       true       true       true\n",
       "     960218     spelta TraesTSP1A TraesCS1A0         1A TraesCS1A0         1A       true       true       true\n",
       "    1080407    stanley TraesSTA1A TraesCS1A0         1A TraesCS1A0         1A       true       true       true\n",
       "    1448033    cadenza TraesCAD_s TraesCS1A0         1A TraesCS1A0 cad_scaffo       true       true       true\n",
       "    1569964     claire TraesCLE_s TraesCS1A0         1A TraesCS1A0 cle_scaffo       true       true       true"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pgsb_df.write_csv(\"pgsb_mapping.csv\")\n",
    "all_genes = pgsb_df.concat ei_mapping\n",
    "all_genes.write_csv(\"all_mapping.csv\")\n",
    "filtered_and_sorted = all_genes.where(all_genes[\"is_unique\"]).sort([\"gene\",\"line\"])\n",
    "filtered_and_sorted.write_csv(\"sorted_filtered_mapping.csv\")\n",
    "filtered_and_sorted.head"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Ruby 2.6.3",
   "language": "ruby",
   "name": "ruby"
  },
  "language_info": {
   "file_extension": ".rb",
   "mimetype": "application/x-ruby",
   "name": "ruby",
   "version": "2.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
