# Filter projected genes 

There are two different sets of gene set. One for the assemblies in pseudomolecules (PGSBv2: ```arinalrfor, chinese, jagger, julius, lancer, landmark, mace, norin61, spelta, stanley, sy_mattis```) and in scaffolds (EI: ```cadenza, claire, paragon, robigus, weebil```). 

## EI annotations

The source gene from the RefSeq v1.1 (IWGSC, 2018) was obtained from ```UK.geneid2CSinformant.txt```, corresponding to the mapping is described in 10+ Wheat Genomes Consortium (2020). Each scaffold was classified according to the corresponding chromosome from the source genes, excluding projections classified as pseudogenes. Then each gene was classified as being consistent or not with the expected scaffold. F


## PGSBv2  annotation

The projected genes where classified as being consistent between the chromosome from the source and projected genes.   

## Both annotations.

Then, the genes on each annotation where filtered to keep the genes that map to the expected chromosome. Finally, the genes that have more than one projection in the expected chromosome where removed.  



```ruby
require 'zlib'
require 'bio'
require 'csv'
require 'set'
require 'bio-gff3'
require 'daru'
require 'nyaplot'
```


    "if(window['d3'] === undefined ||\n   window['Nyaplot'] === undefined){\n    var path = {\"d3\":\"https://cdnjs.cloudflare.com/ajax/libs/d3/3.5.5/d3.min\",\"downloadable\":\"http://cdn.rawgit.com/domitry/d3-downloadable/master/d3-downloadable\"};\n\n\n\n    var shim = {\"d3\":{\"exports\":\"d3\"},\"downloadable\":{\"exports\":\"downloadable\"}};\n\n    require.config({paths: path, shim:shim});\n\n\nrequire(['d3'], function(d3){window['d3']=d3;console.log('finished loading d3');require(['downloadable'], function(downloadable){window['downloadable']=downloadable;console.log('finished loading downloadable');\n\n\tvar script = d3.select(\"head\")\n\t    .append(\"script\")\n\t    .attr(\"src\", \"http://cdn.rawgit.com/domitry/Nyaplotjs/master/release/nyaplot.js\")\n\t    .attr(\"async\", true);\n\n\tscript[0][0].onload = script[0][0].onreadystatechange = function(){\n\n\n\t    var event = document.createEvent(\"HTMLEvents\");\n\t    event.initEvent(\"load_nyaplot\",false,false);\n\t    window.dispatchEvent(event);\n\t    console.log('Finished loading Nyaplotjs');\n\n\t};\n\n\n});});\n}\n"





    false




```ruby
Transcript = Struct.new(:id, :gene, :chromosome,:version,:count,:transcript,:confidence, :count_int, :isoform)
def parseTranscript name
    arr=name.split(".")
    match = /TraesCS(?<chr>[[:alnum:]]{1,2})(?<ver>[[:digit:]]{2})G(?<count>[[:digit:]]+)(?<conf>[[:upper:]]*)/.match arr[0]
    raise "Unable to parse: #{name}" unless match
    Transcript.new(name, arr[0],match[:chr],match[:ver],match[:count],arr[1],match[:conf], match[:count].to_i, arr[1])
end

def parseEITranscript name
    arr=name.split(".")
    match = /Traes(?<chr>[[:upper:]]{3}_scaffold_[[:digit:]]*)_(?<ver>[[:digit:]]{2})G(?<count>[[:digit:]]+)(?<conf>[[:upper:]]*)/.match arr[0]
    raise "Unable to parse: #{name}" unless match
    Transcript.new(name, arr[0],match[:chr].downcase,match[:ver],match[:count],arr[1],match[:conf], match[:count].to_i, arr[1])
end
```




    :parseEITranscript




```ruby
MrnaStats = Struct.new(:cds_count, :cds_max_gap)
module Bio::GFFbrowser::FastLineParser
  module_function :parse_line_fast
end

class GFF3
    def initialize(file: "", is_gz: true)
        @file = file
        @is_gz = is_gz
    end
    
    def header
        return enum_for(:header) unless block_given? 
        io = nil
        if @is_gz
            infile = open(@file)
            io = Zlib::GzipReader.new(infile) 
        else
            io =  File.open(@file)
        end
        
        io.each_line do |line|  
            line.encode!('UTF-8', 'UTF-8', :invalid => :replace)
            line.strip!
            break if line == '##FASTA'
            break unless line.length == 0 or line =~ /^#/
            yield line
        end
    end
    
    def each
        return enum_for(:each) unless block_given? 
        io = nil
        if @is_gz
            infile = open(@file)
            io = Zlib::GzipReader.new(infile) 
        else
            io =  File.open(@file)
        end
        parser = Bio::GFFbrowser::FastLineParser
        io.each_line do |line|  
            line.encode!('UTF-8', 'UTF-8', :invalid => :replace)
            line.strip!
            break if line == '##FASTA'
            #parents.clear if line == '###'
            next if line.length == 0 or line =~ /^#/
            begin
                record = Bio::GFFbrowser::FastLineRecord.new(parser.parse_line_fast(line))
                yield record
            rescue Exception => e
                $stderr.puts "Unable to parse '#{line}'\n#{e}" 
                throw e
            end
         end
    end
    
    def each_gene
        return enum_for(:each_gene) unless block_given? 
        self.each do |record|
           next unless record.feature == "gene"
            yield record
        end 
    end
    
    def each_mrna
        return enum_for(:each_mrna) unless block_given? 
        self.each do |record|
           next unless record.feature == "mRNA"
            yield record
        end 
    end
    
    def each_cds
        return enum_for(:each_mrna) unless block_given? 
        self.each do |record|
           next unless record.feature == "CDS"
            yield record
        end 
    end
    
    def calculate_mrna_stats
        return if @mrna_stats
        
        @mrna_stats = Hash.new {|h,k| h[k] = MrnaStats.new(0,0) }
        
        last_mrna = ""
        last_record = nil
        each_cds do |record|
            parent = record.get_attribute "Parent"
            mrna = @mrna_stats[parent]
            mrna.cds_count += 1
            
            if last_mrna == parent
                distance =  record.start - last_record.end 
                mrna.cds_max_gap = distance if distance > mrna.cds_max_gap
            end
            last_record = record
            last_mrna   = parent

        end
        return
    end
    
    def mrna_info(id)
        calculate_mrna_stats
        @mrna_stats[id] 
    end
    
    def cds_to_print(mrna,cannonical_exons:[])
        cds_features = [] 
        i = 0
        offset=0
        offset_start=0
        each_cds do |record|
            # Target=Sm1_CDS 3049 3183 +
            target = record.get_attribute "Target"
            arr = target.split(" ")
            col = @@colors[i % @@colors.size ]
            start = arr[1].to_i + offset
            ends = arr[2].to_i + offset
            offset_start = record.start  if offset_start == 0
            tmp = CDS_feature.new(start, ends, col, 
                record.seqid, record.start,record.end, record.start - offset_start )
            #offset = ends
            cds_features << tmp
            i += 1
        end
        cds_features
    end
    
end



    
class MultipleGFFs
    def initialize(folder: "../mapping/", lines:[], suffix:".SM1.cds.sorted.gff", is_gz:false )
        @folder = folder
        @lines = lines
        @suffix = suffix
        @lines_gffs = Hash.new
        @lines.each do |l|
            path ="#{folder}/#{l}#{suffix}"
            @lines_gffs[l] = GFF3.new(file: path, is_gz: is_gz)
            #@lines_gffs[l].calculate_mrna_stats
        end
    end
    
    def method_missing(m, *args, &block)
        @lines_gffs.send(m, *args, &block)
    end
    
    def each_gff
       @lines_gffs.each_pair{|k,v| yield k, v }
    end
    
    def summary
        ret = []
        each_gff do |k,v|
            v.each_mrna do |record|
                #coverage=100.0;identity=100.0;matches=4443;mismatches=0;indels=0;unknowns=0
                tmp = {}
                tmp[:line] = k
                tmp[:id] = record.get_attribute "Name"
                tmp[:chr] = record.seqid
                tmp[:start] = record.start
                tmp[:end] = record.end
                tmp[:strand] = record.strand
                tmp[:genomic_length] = record.end - record.start
                tmp[:coverage] = record.get_attribute "coverage"
                tmp[:identity] = record.get_attribute "identity"
                tmp[:matches]  = record.get_attribute "matches"
                tmp[:mismatches]  = record.get_attribute "mismatches"
                tmp[:indels] = record.get_attribute "indels"
                tmp[:unknowns] = record.get_attribute "unknowns"
                mrna_stats = @lines_gffs[k].mrna_info(record.id)
                tmp[:cds_count]   = mrna_stats.cds_count
                tmp[:cds_max_gap] = mrna_stats.cds_max_gap
                ret << tmp
            end
        end
        ret
    end
    
     attr_reader :lines_gffs
    
    def to_svg(mrna: "Sm1_CDS.mrna1", positions: false, out: nil)
        p = Bio::Graphics::Page.new(width: 800,
             height: 1000, 
              number_of_intervals:10,
            background_color: "white"
            )
        each_gff do |k,v|
            generic_track = p.add_track(:glyph => :generic, 
                            :name => k, 
                            :label => true  )
            v.cds_to_print(mrna).each do |cds|
                
                f_id = positions ? cds.offset_start : nil
                feature = Bio::Graphics::MiniFeature.new(start: cds.start, 
                    end: cds.end,
                    fill_color: cds.color, 
                    id: f_id)
                generic_track.add(feature) 
            end 
        end
        
         
        p.write(out) if out
        
        p.get_markup
    end
end
base_path="/Volumes/PanGenome/References"
lines = ["cadenza","claire", "paragon","robigus","weebil"]
gffs = MultipleGFFs.new(folder: "#{base_path}/releaseV3/gff",
    lines:lines, 
    suffix:".gff.gz", 
    is_gz:true )
puts ""

cad = gffs["cadenza"]

valid_genes = Set.new
gffs.each_gff do |k,v|
    v.each_gene do |g|
        valid_genes << g.id
    end
end

valid_genes.size
```

    





    610352



This file contains the source genes for the EI assemblies. 


```ruby
ei_mapping = Daru::DataFrame.from_csv("#{base_path}/releaseV3/gff/UK.geneid2CSinformant.txt", col_sep: "\t")
ei_mapping.size
```




    638111




```ruby
ei_mapping["chromosome"] = ei_mapping.map(:row) do |row|
  parseTranscript(row["transcript"]).chromosome
end

ei_mapping["gene"] = ei_mapping.map(:row) do |row|
  parseTranscript(row["transcript"]).gene
end
ei_mapping["scaffold"] =  ei_mapping.map(:row) do |row|
    parseEITranscript(row["projected_gene"]).chromosome
end
ei_mapping["is_gene"] =  ei_mapping.map(:row) do |row|
    valid_genes.include? row["projected_gene"]
end
ei_mapping.first 10
```




<b> Daru::DataFrame(10x7) </b>

<table border="1" class="dataframe">
  <thead>
    <tr>
      <th></th>
        <th>line</th>
        <th>projected_gene</th>
        <th>transcript</th>
        <th>chromosome</th>
        <th>gene</th>
        <th>scaffold</th>
        <th>is_gene</th>
    </tr>
</thead>
  <tbody>
    <tr>
      <td>0</td>
        <td>cadenza</td>
        <td>TraesCAD_scaffold_000001_01G000100</td>
        <td>TraesCS1B02G148300.1</td>
        <td>1B</td>
        <td>TraesCS1B02G148300</td>
        <td>cad_scaffold_000001</td>
        <td>true</td>
    </tr>
    <tr>
      <td>1</td>
        <td>cadenza</td>
        <td>TraesCAD_scaffold_000001_01G000200</td>
        <td>TraesCSU02G134200.1</td>
        <td>U</td>
        <td>TraesCSU02G134200</td>
        <td>cad_scaffold_000001</td>
        <td>true</td>
    </tr>
    <tr>
      <td>2</td>
        <td>cadenza</td>
        <td>TraesCAD_scaffold_000001_01G000300</td>
        <td>TraesCS7B02G186200.1</td>
        <td>7B</td>
        <td>TraesCS7B02G186200</td>
        <td>cad_scaffold_000001</td>
        <td>true</td>
    </tr>
    <tr>
      <td>3</td>
        <td>cadenza</td>
        <td>TraesCAD_scaffold_000003_01G000100</td>
        <td>TraesCS6D02G266000.1</td>
        <td>6D</td>
        <td>TraesCS6D02G266000</td>
        <td>cad_scaffold_000003</td>
        <td>true</td>
    </tr>
    <tr>
      <td>4</td>
        <td>cadenza</td>
        <td>TraesCAD_scaffold_000003_01G000200</td>
        <td>TraesCS6B02G208700.1</td>
        <td>6B</td>
        <td>TraesCS6B02G208700</td>
        <td>cad_scaffold_000003</td>
      <td>true</td>
</tr><tr>  <td>5</td>
        <td>cadenza</td>
    <td>TraesCAD_scaffold_000003_01G000300</td>
    <td>TraesCS6B02G208800.1</td>
    <td>6B</td>
    <td>TraesCS6B02G208800</td>
    <td>cad_scaffold_000003</td>
    <td>true</td>
</tr><tr>  <td>6</td>
        <td>cadenza</td>
    <td>TraesCAD_scaffold_000004_01G000100</td>
    <td>TraesCS7B02G122100.1</td>
    <td>7B</td>
    <td>TraesCS7B02G122100</td>
    <td>cad_scaffold_000004</td>
    <td>true</td>
</tr><tr>  <td>7</td>
        <td>cadenza</td>
        <td>TraesCAD_scaffold_000004_01G000200</td>
        <td>TraesCS6D02G188400.1</td>
        <td>6D</td>
        <td>TraesCS6D02G188400</td>
        <td>cad_scaffold_000004</td>
        <td>true</td>
    </tr>
    <tr>
      <td>8</td>
        <td>cadenza</td>
        <td>TraesCAD_scaffold_000004_01G000300</td>
        <td>TraesCS5D02G258700.1</td>
        <td>5D</td>
        <td>TraesCS5D02G258700</td>
        <td>cad_scaffold_000004</td>
        <td>true</td>
    </tr>
    <tr>
      <td>9</td>
        <td>cadenza</td>
        <td>TraesCAD_scaffold_000004_01G000400</td>
        <td>TraesCS6B02G357700.1</td>
        <td>6B</td>
        <td>TraesCS6B02G357700</td>
        <td>cad_scaffold_000004</td>
        <td>true</td>
    </tr>
</tbody>
</table>




```ruby
ei_mapping.filter(:row) do |row|
  ["TraesWEE_scaffold_000001_01G000100"].include? row["projected_gene"] 
end
```




<b> Daru::DataFrame(1x7) </b>
<table border="1" class="dataframe">
  <thead>

    <tr>
      <th></th>

        <th>line</th>

        <th>projected_gene</th>

        <th>transcript</th>

        <th>chromosome</th>

        <th>gene</th>

        <th>scaffold</th>

        <th>is_gene</th>

    </tr>

</thead>
  <tbody>

    <tr>
      <td>510301</td>

        <td>weebil</td>

        <td>TraesWEE_scaffold_000001_01G000100</td>

        <td>TraesCS6D02G266000.1</td>

        <td>6D</td>

        <td>TraesCS6D02G266000</td>

        <td>wee_scaffold_000001</td>

        <td>true</td>

    </tr>



</tbody>
</table>




```ruby
ei_mapping_sorted = ei_mapping.sort(["scaffold"])
ei_mapping_sorted.size
```




    638111



The following section takes the sorted table by scaffold to get the chromosome consensus of each scaffold. It is based on the majority of the chromosome of the source gene 


```ruby
i = 0
current_scaffold = ""
current_chromosomes = Hash.new {|h,k| h[k] = 0}
consensus_chromosome = {}
ei_mapping_sorted.each_row do |row|
    if current_scaffold != row["scaffold"]
        if current_chromosomes.size > 0
            max_v = current_chromosomes.max_by{|k,v| v}
            consensus_chromosome[current_scaffold] = [
              current_scaffold, 
              max_v[0], 
              max_v[1],  
              current_chromosomes.values.reduce(:+),
              100 * max_v[1] / current_chromosomes.values.reduce(:+)
            ]
        end
        current_scaffold  = row["scaffold"] 
        current_chromosomes = Hash.new {|h,k| h[k] = 0}
        i+=1
        
    end 
    current_chromosomes[row["chromosome"]] += 1 if row["is_gene"]
    
end
max_v = current_chromosomes.max_by{|k,v| v}

consensus_chromosome[current_scaffold] = [
  current_scaffold, 
  max_v[0],
  max_v[1],
  current_chromosomes.values.reduce(:+), 
  100 * max_v[1] / current_chromosomes.values.reduce(:+)
]

consensus_chr_df = Daru::DataFrame.rows(consensus_chromosome.values , 
    order: [:scaffold, :chr, :hits, :total_genes, :percentage])
consensus_chr_df.head
```




<b> Daru::DataFrame(10x5) </b>
<table border="1" class="dataframe">
  <thead>

    <tr>
      <th></th>

        <th>scaffold</th>

        <th>chr</th>

        <th>hits</th>

        <th>total_genes</th>

        <th>percentage</th>

    </tr>

</thead>
  <tbody>

    <tr>
      <td>0</td>

        <td>cad_scaffold_000001</td>

        <td>1B</td>

        <td>1</td>

        <td>3</td>

        <td>33</td>

    </tr>

    <tr>
      <td>1</td>

        <td>cad_scaffold_000003</td>

        <td>6B</td>

        <td>2</td>

        <td>3</td>

        <td>66</td>

    </tr>

    <tr>
      <td>2</td>

        <td>cad_scaffold_000004</td>

        <td>6B</td>

        <td>2</td>

        <td>5</td>

        <td>40</td>

    </tr>

    <tr>
      <td>3</td>

        <td>cad_scaffold_000005</td>

        <td>7A</td>

        <td>1</td>

        <td>2</td>

        <td>50</td>

    </tr>

    <tr>
      <td>4</td>

        <td>cad_scaffold_000006</td>

        <td>1B</td>

        <td>1</td>

        <td>2</td>

        <td>50</td>

    </tr>

    <tr>
      <td>5</td>

        <td>cad_scaffold_000007</td>

        <td>2B</td>

        <td>4</td>

        <td>6</td>

        <td>66</td>

    </tr>

    <tr>
      <td>6</td>

        <td>cad_scaffold_000008</td>

        <td>7B</td>

        <td>5</td>

        <td>5</td>

        <td>100</td>

    </tr>

    <tr>
      <td>7</td>

        <td>cad_scaffold_000009</td>

        <td>2B</td>

        <td>9</td>

        <td>10</td>

        <td>90</td>

    </tr>

    <tr>
      <td>8</td>

        <td>cad_scaffold_000010</td>

        <td>1B</td>

        <td>2</td>

        <td>2</td>

        <td>100</td>

    </tr>

    <tr>
      <td>9</td>

        <td>cad_scaffold_000011</td>

        <td>7B</td>

        <td>1</td>

        <td>1</td>

        <td>100</td>

    </tr>



</tbody>
</table>




```ruby
consensus_genes_per_scaff = Hash.new {|h,k| h[k] = 0}
genes_per_scaff = Hash.new {|h,k| h[k] = 0}
percentage_cons_per_scaff = Hash.new {|h,k| h[k] = 0}

consensus_chr_df.each_row do |row|
    consensus_genes_per_scaff[  row[:hits]] += 1
    genes_per_scaff[ row[:total_genes]] += 1
    percentage_cons_per_scaff[ 5 * (row[:percentage] / 5 )] += 1
end

percentage_cons_per_scaff
```




    {30=>4156, 65=>12759, 40=>895, 50=>28408, 100=>256295, 90=>313, 80=>4506, 45=>14, 75=>6094, 60=>1879, 25=>731, 70=>624, 55=>398, 85=>1420, 35=>69, 95=>4, 20=>108, 15=>14, 10=>1}




```ruby
percentage_cons_df = Daru::DataFrame.new({
  percentage: percentage_cons_per_scaff.keys,
  count: percentage_cons_per_scaff.values,
  })
percentage_cons_df.sort!([:percentage])
plot = Nyaplot::Plot.new
bar = plot.add(:line, percentage_cons_df[:percentage].to_a, percentage_cons_df[:count].to_a)
plot.x_label "Percentage of genes corresponding to consensus"
plot.y_label "Count"
plot.show
puts ""
```


<div id='vis-2df2e404-24d3-4e4a-acfb-739db8466683'></div>
<script>
(function(){
    var render = function(){
        var model = {"panes":[{"diagrams":[{"type":"line","options":{"x":"data0","y":"data1"},"data":"1c3e99c5-88a4-4c96-afcc-341ab367dd9e"}],"options":{"x_label":"Percentage of genes corresponding to consensus","y_label":"Count","zoom":true,"width":700,"xrange":[10,100],"yrange":[1,256295]}}],"data":{"1c3e99c5-88a4-4c96-afcc-341ab367dd9e":[{"data0":10,"data1":1},{"data0":15,"data1":14},{"data0":20,"data1":108},{"data0":25,"data1":731},{"data0":30,"data1":4156},{"data0":35,"data1":69},{"data0":40,"data1":895},{"data0":45,"data1":14},{"data0":50,"data1":28408},{"data0":55,"data1":398},{"data0":60,"data1":1879},{"data0":65,"data1":12759},{"data0":70,"data1":624},{"data0":75,"data1":6094},{"data0":80,"data1":4506},{"data0":85,"data1":1420},{"data0":90,"data1":313},{"data0":95,"data1":4},{"data0":100,"data1":256295}]},"extension":[]}
        var id_name = '#vis-2df2e404-24d3-4e4a-acfb-739db8466683';
        Nyaplot.core.parse(model, id_name);

        require(['downloadable'], function(downloadable){
          var svg = d3.select(id_name).select("svg");
	  if(!svg.empty())
	    svg.call(downloadable().filename('fig'));
	});
    };
    if(window['Nyaplot']==undefined){
        window.addEventListener('load_nyaplot', render, false);
	return;
    } else {
       render();
    }
})();
</script>



    



```ruby
consensus_chr_df.write_csv("./EI_consensus_genes.csv")
```


```ruby
cons_hash = {}
consensus_chr_df.each_row{|row| cons_hash[row[:scaffold]] = row[:chr]}
cons_hash.first

```




    ["cad_scaffold_000001", "1B"]



Comparing each consensus chromosome to the gene.


```ruby
ei_mapping["is_chr_gene_consistent"] =  ei_mapping.map(:row) do |row|
    cons_hash[row["scaffold"]] == row["chromosome"]
end
ei_mapping.first 20
```




<b> Daru::DataFrame(20x8) </b>
<table border="1" class="dataframe">
  <thead>

    <tr>
      <th></th>

        <th>line</th>

        <th>projected_gene</th>

        <th>transcript</th>

        <th>chromosome</th>

        <th>gene</th>

        <th>scaffold</th>

        <th>is_gene</th>

        <th>is_chr_gene_consistent</th>

    </tr>

</thead>
  <tbody>

    <tr>
      <td>0</td>

        <td>cadenza</td>

        <td>TraesCAD_scaffold_000001_01G000100</td>

        <td>TraesCS1B02G148300.1</td>

        <td>1B</td>

        <td>TraesCS1B02G148300</td>

        <td>cad_scaffold_000001</td>

        <td>true</td>

        <td>true</td>

    </tr>

    <tr>
      <td>1</td>

        <td>cadenza</td>

        <td>TraesCAD_scaffold_000001_01G000200</td>

        <td>TraesCSU02G134200.1</td>

        <td>U</td>

        <td>TraesCSU02G134200</td>

        <td>cad_scaffold_000001</td>

        <td>true</td>

        <td>false</td>

    </tr>

    <tr>
      <td>2</td>

        <td>cadenza</td>

        <td>TraesCAD_scaffold_000001_01G000300</td>

        <td>TraesCS7B02G186200.1</td>

        <td>7B</td>

        <td>TraesCS7B02G186200</td>

        <td>cad_scaffold_000001</td>

        <td>true</td>

        <td>false</td>

    </tr>

    <tr>
      <td>3</td>

        <td>cadenza</td>

        <td>TraesCAD_scaffold_000003_01G000100</td>

        <td>TraesCS6D02G266000.1</td>

        <td>6D</td>

        <td>TraesCS6D02G266000</td>

        <td>cad_scaffold_000003</td>

        <td>true</td>

        <td>false</td>

    </tr>

    <tr>
      <td>4</td>

        <td>cadenza</td>

        <td>TraesCAD_scaffold_000003_01G000200</td>

        <td>TraesCS6B02G208700.1</td>

        <td>6B</td>

        <td>TraesCS6B02G208700</td>

        <td>cad_scaffold_000003</td>

        <td>true</td>

        <td>true</td>

    </tr>

    <tr>
      <td>5</td>

        <td>cadenza</td>

        <td>TraesCAD_scaffold_000003_01G000300</td>

        <td>TraesCS6B02G208800.1</td>

        <td>6B</td>

        <td>TraesCS6B02G208800</td>

        <td>cad_scaffold_000003</td>

        <td>true</td>

        <td>true</td>

    </tr>

    <tr>
      <td>6</td>

        <td>cadenza</td>

        <td>TraesCAD_scaffold_000004_01G000100</td>

        <td>TraesCS7B02G122100.1</td>

        <td>7B</td>

        <td>TraesCS7B02G122100</td>

        <td>cad_scaffold_000004</td>

        <td>true</td>

        <td>false</td>

    </tr>

    <tr>
      <td>7</td>

        <td>cadenza</td>

        <td>TraesCAD_scaffold_000004_01G000200</td>

        <td>TraesCS6D02G188400.1</td>

        <td>6D</td>

        <td>TraesCS6D02G188400</td>

        <td>cad_scaffold_000004</td>

        <td>true</td>

        <td>false</td>

    </tr>

    <tr>
      <td>8</td>

        <td>cadenza</td>

        <td>TraesCAD_scaffold_000004_01G000300</td>

        <td>TraesCS5D02G258700.1</td>

        <td>5D</td>

        <td>TraesCS5D02G258700</td>

        <td>cad_scaffold_000004</td>

        <td>true</td>

        <td>false</td>

    </tr>

    <tr>
      <td>9</td>

        <td>cadenza</td>

        <td>TraesCAD_scaffold_000004_01G000400</td>

        <td>TraesCS6B02G357700.1</td>

        <td>6B</td>

        <td>TraesCS6B02G357700</td>

        <td>cad_scaffold_000004</td>

        <td>true</td>

        <td>true</td>

    </tr>

    <tr>
      <td>10</td>

        <td>cadenza</td>

        <td>TraesCAD_scaffold_000004_01G000500</td>

        <td>TraesCS6B02G357800.1</td>

        <td>6B</td>

        <td>TraesCS6B02G357800</td>

        <td>cad_scaffold_000004</td>

        <td>true</td>

        <td>true</td>

    </tr>

    <tr>
      <td>11</td>

        <td>cadenza</td>

        <td>TraesCAD_scaffold_000005_01G000100</td>

        <td>TraesCS7A02G427500.1</td>

        <td>7A</td>

        <td>TraesCS7A02G427500</td>

        <td>cad_scaffold_000005</td>

        <td>true</td>

        <td>true</td>

    </tr>

    <tr>
      <td>12</td>

        <td>cadenza</td>

        <td>TraesCAD_scaffold_000005_01G000200</td>

        <td>TraesCS4A02G150700.1</td>

        <td>4A</td>

        <td>TraesCS4A02G150700</td>

        <td>cad_scaffold_000005</td>

        <td>true</td>

        <td>false</td>

    </tr>

    <tr>
      <td>13</td>

        <td>cadenza</td>

        <td>TraesCAD_scaffold_000006_01G000100</td>

        <td>TraesCS1B02G148400.1</td>

        <td>1B</td>

        <td>TraesCS1B02G148400</td>

        <td>cad_scaffold_000006</td>

        <td>true</td>

        <td>true</td>

    </tr>

    <tr>
      <td>14</td>

        <td>cadenza</td>

        <td>TraesCAD_scaffold_000006_01G000200</td>

        <td>TraesCS6B02G063600.1</td>

        <td>6B</td>

        <td>TraesCS6B02G063600</td>

        <td>cad_scaffold_000006</td>

        <td>true</td>

        <td>false</td>

    </tr>

    <tr>
      <td>15</td>

        <td>cadenza</td>

        <td>TraesCAD_scaffold_000007_01G000100</td>

        <td>TraesCS4B02G081300.1</td>

        <td>4B</td>

        <td>TraesCS4B02G081300</td>

        <td>cad_scaffold_000007</td>

        <td>true</td>

        <td>false</td>

    </tr>

    <tr>
      <td>16</td>

        <td>cadenza</td>

        <td>TraesCAD_scaffold_000007_01G000200</td>

        <td>TraesCS2B02G399100.1</td>

        <td>2B</td>

        <td>TraesCS2B02G399100</td>

        <td>cad_scaffold_000007</td>

        <td>true</td>

        <td>true</td>

    </tr>

    <tr>
      <td>17</td>

        <td>cadenza</td>

        <td>TraesCAD_scaffold_000007_01G000300</td>

        <td>TraesCS2B02G399000.1</td>

        <td>2B</td>

        <td>TraesCS2B02G399000</td>

        <td>cad_scaffold_000007</td>

        <td>true</td>

        <td>true</td>

    </tr>

    <tr>
      <td>18</td>

        <td>cadenza</td>

        <td>TraesCAD_scaffold_000007_01G000400</td>

        <td>TraesCS2B02G398900.1</td>

        <td>2B</td>

        <td>TraesCS2B02G398900</td>

        <td>cad_scaffold_000007</td>

        <td>true</td>

        <td>true</td>

    </tr>

    <tr>
      <td>19</td>

        <td>cadenza</td>

        <td>TraesCAD_scaffold_000007_01G000500</td>

        <td>TraesCS2D02G466900.1</td>

        <td>2D</td>

        <td>TraesCS2D02G466900</td>

        <td>cad_scaffold_000007</td>

        <td>true</td>

        <td>false</td>

    </tr>



</tbody>
</table>



Comparing that it is not a pseudogene and that it is consistent. 


```ruby
gene_count = Hash.new{|h,k| h[k] = 0}
ei_mapping.each_row do |row|
    next unless row["is_gene"] and row["is_chr_gene_consistent"]
    gene_count[row["line"] + "_" + row["gene"]] += 1
end
puts gene_count.size
```

    455268


Finally, comparing that the gene is mapped only once (Removes posible dupplications or genes that are ambigous)


```ruby
ei_mapping["is_unique"] =  ei_mapping.map(:row) do |row|
    gene_count[row["line"] + "_" + row["gene"]] == 1 and  row["is_gene"] and row["is_chr_gene_consistent"] 
end
ei_mapping.head
```




<b> Daru::DataFrame(10x9) </b>
<table border="1" class="dataframe">
  <thead>

    <tr>
      <th></th>

        <th>line</th>

        <th>projected_gene</th>

        <th>transcript</th>

        <th>chromosome</th>

        <th>gene</th>

        <th>scaffold</th>

        <th>is_gene</th>

        <th>is_chr_gene_consistent</th>

        <th>is_unique</th>

    </tr>

</thead>
  <tbody>

    <tr>
      <td>0</td>

        <td>cadenza</td>

        <td>TraesCAD_scaffold_000001_01G000100</td>

        <td>TraesCS1B02G148300.1</td>

        <td>1B</td>

        <td>TraesCS1B02G148300</td>

        <td>cad_scaffold_000001</td>

        <td>true</td>

        <td>true</td>

        <td>false</td>

    </tr>

    <tr>
      <td>1</td>

        <td>cadenza</td>

        <td>TraesCAD_scaffold_000001_01G000200</td>

        <td>TraesCSU02G134200.1</td>

        <td>U</td>

        <td>TraesCSU02G134200</td>

        <td>cad_scaffold_000001</td>

        <td>true</td>

        <td>false</td>

        <td>false</td>

    </tr>

    <tr>
      <td>2</td>

        <td>cadenza</td>

        <td>TraesCAD_scaffold_000001_01G000300</td>

        <td>TraesCS7B02G186200.1</td>

        <td>7B</td>

        <td>TraesCS7B02G186200</td>

        <td>cad_scaffold_000001</td>

        <td>true</td>

        <td>false</td>

        <td>false</td>

    </tr>

    <tr>
      <td>3</td>

        <td>cadenza</td>

        <td>TraesCAD_scaffold_000003_01G000100</td>

        <td>TraesCS6D02G266000.1</td>

        <td>6D</td>

        <td>TraesCS6D02G266000</td>

        <td>cad_scaffold_000003</td>

        <td>true</td>

        <td>false</td>

        <td>false</td>

    </tr>

    <tr>
      <td>4</td>

        <td>cadenza</td>

        <td>TraesCAD_scaffold_000003_01G000200</td>

        <td>TraesCS6B02G208700.1</td>

        <td>6B</td>

        <td>TraesCS6B02G208700</td>

        <td>cad_scaffold_000003</td>

        <td>true</td>

        <td>true</td>

        <td>false</td>

    </tr>

    <tr>
      <td>5</td>

        <td>cadenza</td>

        <td>TraesCAD_scaffold_000003_01G000300</td>

        <td>TraesCS6B02G208800.1</td>

        <td>6B</td>

        <td>TraesCS6B02G208800</td>

        <td>cad_scaffold_000003</td>

        <td>true</td>

        <td>true</td>

        <td>true</td>

    </tr>

    <tr>
      <td>6</td>

        <td>cadenza</td>

        <td>TraesCAD_scaffold_000004_01G000100</td>

        <td>TraesCS7B02G122100.1</td>

        <td>7B</td>

        <td>TraesCS7B02G122100</td>

        <td>cad_scaffold_000004</td>

        <td>true</td>

        <td>false</td>

        <td>false</td>

    </tr>

    <tr>
      <td>7</td>

        <td>cadenza</td>

        <td>TraesCAD_scaffold_000004_01G000200</td>

        <td>TraesCS6D02G188400.1</td>

        <td>6D</td>

        <td>TraesCS6D02G188400</td>

        <td>cad_scaffold_000004</td>

        <td>true</td>

        <td>false</td>

        <td>false</td>

    </tr>

    <tr>
      <td>8</td>

        <td>cadenza</td>

        <td>TraesCAD_scaffold_000004_01G000300</td>

        <td>TraesCS5D02G258700.1</td>

        <td>5D</td>

        <td>TraesCS5D02G258700</td>

        <td>cad_scaffold_000004</td>

        <td>true</td>

        <td>false</td>

        <td>false</td>

    </tr>

    <tr>
      <td>9</td>

        <td>cadenza</td>

        <td>TraesCAD_scaffold_000004_01G000400</td>

        <td>TraesCS6B02G357700.1</td>

        <td>6B</td>

        <td>TraesCS6B02G357700</td>

        <td>cad_scaffold_000004</td>

        <td>true</td>

        <td>true</td>

        <td>true</td>

    </tr>



</tbody>
</table>




```ruby
consensus_chr_df.write_csv("EI_consensus_genes.csv")
ei_mapping.write_csv("ei_mapping.csv")
```


```ruby
ei_mapping["line"].uniq
```




<b> Daru::Vector(5) </b>
<table>
  <thead>

    <tr>
        <th> </th>
        <th>line</th>
    </tr>

</thead>
  <tbody>

    <tr>
      <td>0</td>
      <td>cadenza</td>
    </tr>

    <tr>
      <td>128544</td>
      <td>claire</td>
    </tr>

    <tr>
      <td>255850</td>
      <td>paragon</td>
    </tr>

    <tr>
      <td>383153</td>
      <td>robigus</td>
    </tr>

    <tr>
      <td>510301</td>
      <td>weebil</td>
    </tr>



</tbody>
</table>



# PGSB Assemblies 


```ruby
lines = [
    "arinalrfor",
    "chinese",
    "jagger",
    "julius",
    "lancer",
    "landmark",
    "mace",
    "norin61",
    "spelta",
    "stanley",
    "sy_mattis"
    ]
gffs_pgsb = MultipleGFFs.new(folder: "#{base_path}/releasePGSBv2.0/gff",
    lines:lines, 
    suffix:".gff.gz", 
    is_gz:true )
puts ""
```

    



```ruby
def parsePGSBTranscript name
    arr=name.split(".")
    match = /Traes(?<variety>[[:upper:]]{3})(?<chr>[[:alnum:]]{1,2})(?<ver>[[:digit:]]{2})G(?<count>[[:digit:]]+)(?<conf>[[:upper:]]*)/.match arr[0]
    #match  = /Traes(?<variety>[[:upper:]]*)/.match arr[0]
    #puts match.inspect
    raise "Unable to parse: #{name}" unless match
    Transcript.new(name, arr[0],match[:chr],match[:ver],match[:count],match[:variety], match[:conf],match[:count].to_i, arr[1])
end
rows = []
gffs_pgsb.each_gff do |k,v|
    v.each_gene do |record|
        srcmodel = record.attributes["srcmodel"]
        parsed_proj = parsePGSBTranscript record.id
        parsed_src = parseTranscript srcmodel
        
        rows << [k, 
            record.id, 
            record.attributes["srcmodel"],
            parsed_src.chromosome, 
            parsed_src.gene,
            parsed_proj.chromosome,  
            record.attributes["disruptedORF"].nil?, 
            parsed_proj.chromosome == parsed_src.chromosome ]      
    end
end

pgsb_df = Daru::DataFrame.rows(rows , 
    order: ["line", "projected_gene", "transcript", "chromosome", "gene", "scaffold", "is_gene", "is_chr_gene_consistent"])
pgsb_df.head
```




<b> Daru::DataFrame(10x8) </b>
<table border="1" class="dataframe">
  <thead>

    <tr>
      <th></th>

        <th>line</th>

        <th>projected_gene</th>

        <th>transcript</th>

        <th>chromosome</th>

        <th>gene</th>

        <th>scaffold</th>

        <th>is_gene</th>

        <th>is_chr_gene_consistent</th>

    </tr>

</thead>
  <tbody>

    <tr>
      <td>0</td>

        <td>arinalrfor</td>

        <td>TraesARI1A01G000100</td>

        <td>TraesCS1A02G003546.1</td>

        <td>1A</td>

        <td>TraesCS1A02G003546</td>

        <td>1A</td>

        <td>false</td>

        <td>true</td>

    </tr>

    <tr>
      <td>1</td>

        <td>arinalrfor</td>

        <td>TraesARI1A01G000200</td>

        <td>TraesCS1A02G004126.1</td>

        <td>1A</td>

        <td>TraesCS1A02G004126</td>

        <td>1A</td>

        <td>true</td>

        <td>true</td>

    </tr>

    <tr>
      <td>2</td>

        <td>arinalrfor</td>

        <td>TraesARI1A01G000300</td>

        <td>TraesCS1A02G003600.1</td>

        <td>1A</td>

        <td>TraesCS1A02G003600</td>

        <td>1A</td>

        <td>true</td>

        <td>true</td>

    </tr>

    <tr>
      <td>3</td>

        <td>arinalrfor</td>

        <td>TraesARI1A01G000400</td>

        <td>TraesCS1A02G003700.1</td>

        <td>1A</td>

        <td>TraesCS1A02G003700</td>

        <td>1A</td>

        <td>true</td>

        <td>true</td>

    </tr>

    <tr>
      <td>4</td>

        <td>arinalrfor</td>

        <td>TraesARI1A01G000500</td>

        <td>TraesCS7A02G485500.1</td>

        <td>7A</td>

        <td>TraesCS7A02G485500</td>

        <td>1A</td>

        <td>true</td>

        <td>false</td>

    </tr>

    <tr>
      <td>5</td>

        <td>arinalrfor</td>

        <td>TraesARI1A01G000600</td>

        <td>TraesCS3A02G080300.1</td>

        <td>3A</td>

        <td>TraesCS3A02G080300</td>

        <td>1A</td>

        <td>true</td>

        <td>false</td>

    </tr>

    <tr>
      <td>6</td>

        <td>arinalrfor</td>

        <td>TraesARI1A01G000700</td>

        <td>TraesCS6D02G375000.1</td>

        <td>6D</td>

        <td>TraesCS6D02G375000</td>

        <td>1A</td>

        <td>true</td>

        <td>false</td>

    </tr>

    <tr>
      <td>7</td>

        <td>arinalrfor</td>

        <td>TraesARI1A01G000800</td>

        <td>TraesCSU02G097400.1</td>

        <td>U</td>

        <td>TraesCSU02G097400</td>

        <td>1A</td>

        <td>true</td>

        <td>false</td>

    </tr>

    <tr>
      <td>8</td>

        <td>arinalrfor</td>

        <td>TraesARI1A01G000900</td>

        <td>TraesCS1A02G019300.1</td>

        <td>1A</td>

        <td>TraesCS1A02G019300</td>

        <td>1A</td>

        <td>true</td>

        <td>true</td>

    </tr>

    <tr>
      <td>9</td>

        <td>arinalrfor</td>

        <td>TraesARI1A01G001000</td>

        <td>TraesCS1A02G003500.1</td>

        <td>1A</td>

        <td>TraesCS1A02G003500</td>

        <td>1A</td>

        <td>true</td>

        <td>true</td>

    </tr>



</tbody>
</table>



Validating that the chromosome from the source gene is consistent with the chromosome. 


```ruby
gene_count_pgsb = Hash.new{|h,k| h[k] = 0}
pgsb_df.each_row do |row|
    next unless row["is_gene"] and row["is_chr_gene_consistent"]
    gene_count_pgsb[row["line"] + "_" + row["gene"]] += 1
end
puts gene_count_pgsb.size
```

    1069393



```ruby
pgsb_df["is_unique"] =  pgsb_df.map(:row) do |row|
    gene_count_pgsb[row["line"] + "_" + row["gene"]] == 1 and  row["is_gene"] and row["is_chr_gene_consistent"] 
end
pgsb_df.head
```




<b> Daru::DataFrame(10x9) </b>
<table border="1" class="dataframe">
  <thead>

    <tr>
      <th></th>

        <th>line</th>

        <th>projected_gene</th>

        <th>transcript</th>

        <th>chromosome</th>

        <th>gene</th>

        <th>scaffold</th>

        <th>is_gene</th>

        <th>is_chr_gene_consistent</th>

        <th>is_unique</th>

    </tr>

</thead>
  <tbody>

    <tr>
      <td>0</td>

        <td>arinalrfor</td>

        <td>TraesARI1A01G000100</td>

        <td>TraesCS1A02G003546.1</td>

        <td>1A</td>

        <td>TraesCS1A02G003546</td>

        <td>1A</td>

        <td>false</td>

        <td>true</td>

        <td>false</td>

    </tr>

    <tr>
      <td>1</td>

        <td>arinalrfor</td>

        <td>TraesARI1A01G000200</td>

        <td>TraesCS1A02G004126.1</td>

        <td>1A</td>

        <td>TraesCS1A02G004126</td>

        <td>1A</td>

        <td>true</td>

        <td>true</td>

        <td>true</td>

    </tr>

    <tr>
      <td>2</td>

        <td>arinalrfor</td>

        <td>TraesARI1A01G000300</td>

        <td>TraesCS1A02G003600.1</td>

        <td>1A</td>

        <td>TraesCS1A02G003600</td>

        <td>1A</td>

        <td>true</td>

        <td>true</td>

        <td>true</td>

    </tr>

    <tr>
      <td>3</td>

        <td>arinalrfor</td>

        <td>TraesARI1A01G000400</td>

        <td>TraesCS1A02G003700.1</td>

        <td>1A</td>

        <td>TraesCS1A02G003700</td>

        <td>1A</td>

        <td>true</td>

        <td>true</td>

        <td>true</td>

    </tr>

    <tr>
      <td>4</td>

        <td>arinalrfor</td>

        <td>TraesARI1A01G000500</td>

        <td>TraesCS7A02G485500.1</td>

        <td>7A</td>

        <td>TraesCS7A02G485500</td>

        <td>1A</td>

        <td>true</td>

        <td>false</td>

        <td>false</td>

    </tr>

    <tr>
      <td>5</td>

        <td>arinalrfor</td>

        <td>TraesARI1A01G000600</td>

        <td>TraesCS3A02G080300.1</td>

        <td>3A</td>

        <td>TraesCS3A02G080300</td>

        <td>1A</td>

        <td>true</td>

        <td>false</td>

        <td>false</td>

    </tr>

    <tr>
      <td>6</td>

        <td>arinalrfor</td>

        <td>TraesARI1A01G000700</td>

        <td>TraesCS6D02G375000.1</td>

        <td>6D</td>

        <td>TraesCS6D02G375000</td>

        <td>1A</td>

        <td>true</td>

        <td>false</td>

        <td>false</td>

    </tr>

    <tr>
      <td>7</td>

        <td>arinalrfor</td>

        <td>TraesARI1A01G000800</td>

        <td>TraesCSU02G097400.1</td>

        <td>U</td>

        <td>TraesCSU02G097400</td>

        <td>1A</td>

        <td>true</td>

        <td>false</td>

        <td>false</td>

    </tr>

    <tr>
      <td>8</td>

        <td>arinalrfor</td>

        <td>TraesARI1A01G000900</td>

        <td>TraesCS1A02G019300.1</td>

        <td>1A</td>

        <td>TraesCS1A02G019300</td>

        <td>1A</td>

        <td>true</td>

        <td>true</td>

        <td>true</td>

    </tr>

    <tr>
      <td>9</td>

        <td>arinalrfor</td>

        <td>TraesARI1A01G001000</td>

        <td>TraesCS1A02G003500.1</td>

        <td>1A</td>

        <td>TraesCS1A02G003500</td>

        <td>1A</td>

        <td>true</td>

        <td>true</td>

        <td>true</td>

    </tr>



</tbody>
</table>




```ruby
pgsb_df.write_csv("pgsb_mapping.csv")
all_genes = pgsb_df.concat ei_mapping
all_genes.write_csv("all_mapping.csv")
filtered_and_sorted = all_genes.where(all_genes["is_unique"]).sort(["gene","line"])
filtered_and_sorted.write_csv("sorted_filtered_mapping.csv")
filtered_and_sorted.head
```




<b> Daru::DataFrame(10x9) </b>
<table border="1" class="dataframe">
  <thead>

    <tr>
      <th></th>

        <th>line</th>

        <th>projected_gene</th>

        <th>transcript</th>

        <th>chromosome</th>

        <th>gene</th>

        <th>scaffold</th>

        <th>is_gene</th>

        <th>is_chr_gene_consistent</th>

        <th>is_unique</th>

    </tr>

</thead>
  <tbody>

    <tr>
      <td>244046</td>

        <td>jagger</td>

        <td>TraesJAG1A01G000500</td>

        <td>TraesCS1A02G000100.1</td>

        <td>1A</td>

        <td>TraesCS1A02G000100</td>

        <td>1A</td>

        <td>true</td>

        <td>true</td>

        <td>true</td>

    </tr>

    <tr>
      <td>363868</td>

        <td>julius</td>

        <td>TraesJUL1A01G036600</td>

        <td>TraesCS1A02G000100.1</td>

        <td>1A</td>

        <td>TraesCS1A02G000100</td>

        <td>1A</td>

        <td>true</td>

        <td>true</td>

        <td>true</td>

    </tr>

    <tr>
      <td>482640</td>

        <td>lancer</td>

        <td>TraesLAC1A01G000300</td>

        <td>TraesCS1A02G000100.1</td>

        <td>1A</td>

        <td>TraesCS1A02G000100</td>

        <td>1A</td>

        <td>true</td>

        <td>true</td>

        <td>true</td>

    </tr>

    <tr>
      <td>602686</td>

        <td>landmark</td>

        <td>TraesLDM1A01G000400</td>

        <td>TraesCS1A02G000100.1</td>

        <td>1A</td>

        <td>TraesCS1A02G000100</td>

        <td>1A</td>

        <td>true</td>

        <td>true</td>

        <td>true</td>

    </tr>

    <tr>
      <td>721733</td>

        <td>mace</td>

        <td>TraesMAC1A01G002400</td>

        <td>TraesCS1A02G000100.1</td>

        <td>1A</td>

        <td>TraesCS1A02G000100</td>

        <td>1A</td>

        <td>true</td>

        <td>true</td>

        <td>true</td>

    </tr>

    <tr>
      <td>841525</td>

        <td>norin61</td>

        <td>TraesNOR1A01G004400</td>

        <td>TraesCS1A02G000100.1</td>

        <td>1A</td>

        <td>TraesCS1A02G000100</td>

        <td>1A</td>

        <td>true</td>

        <td>true</td>

        <td>true</td>

    </tr>

    <tr>
      <td>960218</td>

        <td>spelta</td>

        <td>TraesTSP1A01G000300</td>

        <td>TraesCS1A02G000100.1</td>

        <td>1A</td>

        <td>TraesCS1A02G000100</td>

        <td>1A</td>

        <td>true</td>

        <td>true</td>

        <td>true</td>

    </tr>

    <tr>
      <td>1080407</td>

        <td>stanley</td>

        <td>TraesSTA1A01G008800</td>

        <td>TraesCS1A02G000100.1</td>

        <td>1A</td>

        <td>TraesCS1A02G000100</td>

        <td>1A</td>

        <td>true</td>

        <td>true</td>

        <td>true</td>

    </tr>

    <tr>
      <td>1448033</td>

        <td>cadenza</td>

        <td>TraesCAD_scaffold_628837_01G000200</td>

        <td>TraesCS1A02G000200.1</td>

        <td>1A</td>

        <td>TraesCS1A02G000200</td>

        <td>cad_scaffold_628837</td>

        <td>true</td>

        <td>true</td>

        <td>true</td>

    </tr>

    <tr>
      <td>1569964</td>

        <td>claire</td>

        <td>TraesCLE_scaffold_182945_01G000100</td>

        <td>TraesCS1A02G000200.1</td>

        <td>1A</td>

        <td>TraesCS1A02G000200</td>

        <td>cle_scaffold_182945</td>

        <td>true</td>

        <td>true</td>

        <td>true</td>

    </tr>



</tbody>
</table>


